{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7fc78df8-ed85-43d9-9428-512929c8868d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# License: GNU Affero General Public License v3 or later\n",
    "# A copy of GNU AGPL v3 should have been included in this software package in LICENSE.txt.\n",
    "\n",
    "\"\"\" A collection of functions for reading, processing, and sanitising records.\n",
    "\"\"\"\n",
    "\n",
    "import functools\n",
    "import logging\n",
    "import re\n",
    "from typing import Any, Dict, Callable, List, Set, Tuple\n",
    "import warnings\n",
    "\n",
    "from Bio.Seq import Seq, UnknownSeq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from helperlibs.bio import seqio\n",
    "\n",
    "#from antismash.common import gff_parser\n",
    "from antismash.common.errors import AntismashInputError\n",
    "from antismash.common.secmet import Record\n",
    "from antismash.common.secmet.errors import SecmetInvalidInputError\n",
    "from antismash.config import get_config, update_config, Config, ConfigType\n",
    "from antismash.custom_typing import AntismashModule\n",
    "\n",
    "#from subprocessing import parallel_function\n",
    "\n",
    "\n",
    "def _strict_parse(filename: str) -> List[SeqRecord]:\n",
    "    \"\"\" Parses the input record with extra wrappers to catch biopython warnings\n",
    "        as errors.\n",
    "\n",
    "        Arguments:\n",
    "            filename: the name of the file to parse\n",
    "\n",
    "        Returns:\n",
    "            a list of SeqRecords parsed\n",
    "    \"\"\"\n",
    "    filter_messages = [\n",
    "        r\".*invalid location.*\",\n",
    "        r\".*Expected sequence length.*\",\n",
    "        r\".*Couldn't parse feature location.*\",\n",
    "        r\".*double-quote characters.*should be escaped.*\",\n",
    "    ]\n",
    "    try:\n",
    "        # prepend warning filters to raise exceptions on certain messages\n",
    "        for message in filter_messages:\n",
    "            warnings.filterwarnings(\"error\", message=message)\n",
    "        records = list(seqio.parse(filename))\n",
    "    except Exception as err:\n",
    "        message = str(err)\n",
    "        # strip the \"Ignoring\" part, since it's not being ignored\n",
    "        if message.startswith(\"Ignoring invalid location\"):\n",
    "            message = message[9:]\n",
    "        logging.error('Parsing %r failed: %s', filename, message)\n",
    "        raise AntismashInputError(message) from err\n",
    "    finally:\n",
    "        # remove the new warning filters (functions in at least 3.5 and 3.6)\n",
    "        # since mypy doesn't recognise this attribute, ignore the type\n",
    "        warnings.filters = warnings.filters[len(filter_messages):]   # type: ignore\n",
    "\n",
    "    if not records:\n",
    "        raise AntismashInputError(\"no valid records found in file %s\" % filename)\n",
    "    return records\n",
    "\n",
    "\n",
    "def parse_input_sequence(filename: str, taxon: str = \"bacteria\", minimum_length: int = -1,\n",
    "                         start: int = -1, end: int = -1, gff_file: str = \"\") -> List[Record]:\n",
    "    \"\"\" Parse input records contained in a file\n",
    "\n",
    "        Arguments:\n",
    "            filename: the path of the file to read\n",
    "            taxon: the taxon of the input, e.g. 'bacteria', 'fungi'\n",
    "            minimum_length: records with length less than this will be ignored\n",
    "                            if not positive, all records are included\n",
    "            start: a start location for trimming the sequence, or -1 to use all\n",
    "            end: an end location for trimming the sequence, or -1 to use all\n",
    "            gff_file: a GFF file to use for gene/CDS annotations\n",
    "\n",
    "        Returns:\n",
    "            A list of secmet.Record instances, one for each record in the file\n",
    "    \"\"\"\n",
    "    logging.info('Parsing input sequence %r', filename)\n",
    "    if not isinstance(minimum_length, int):\n",
    "        raise TypeError(\"minimum_length must be an int\")\n",
    "\n",
    "    records: List[SeqRecord] = []\n",
    "\n",
    "    for record in _strict_parse(filename):\n",
    "        if minimum_length < 1 \\\n",
    "                or len(record.seq) >= minimum_length \\\n",
    "                or 'contig' in record.annotations \\\n",
    "                or 'wgs_scafld' in record.annotations \\\n",
    "                or 'wgs' in record.annotations:\n",
    "            records.append(record)\n",
    "\n",
    "    # if no records are left, that's a problem\n",
    "    if not records:\n",
    "        raise AntismashInputError(\"all input records smaller than minimum length (%d)\" % minimum_length)\n",
    "\n",
    "    for record in records:\n",
    "        if not Record.is_nucleotide_sequence(record.seq):\n",
    "            raise AntismashInputError(\"protein records are not supported: %s\" % record.id)\n",
    "\n",
    "    # before conversion to secmet records, trim if required\n",
    "    if start > -1 or end > -1:\n",
    "        if len(records) > 1:\n",
    "            raise ValueError(\"--start and --end options cannot be used with multiple records\")\n",
    "        records[0] = trim_sequence(records[0], max(start, 0), min(len(records[0]), end))\n",
    "\n",
    "    # add GFF features before conversion, if relevant\n",
    "    if gff_file:\n",
    "        logging.debug(\"Loading annotations from GFF file\")\n",
    "        # check GFF suitability first\n",
    "        try:\n",
    "            check_gff_suitability(gff_file, records)\n",
    "        except AntismashInputError:\n",
    "            raise\n",
    "        except Exception as err:\n",
    "            # avoid swallowing details if possible\n",
    "            if str(err):\n",
    "                logging.error(err)\n",
    "            raise AntismashInputError(\"could not parse records from GFF3 file\") from err\n",
    "        gff_features = run(gff_file)\n",
    "        for record in records:\n",
    "            if any(feature.type == \"CDS\" for feature in record.features):\n",
    "                continue\n",
    "            record.features.extend(gff_features.get(record.id, []))\n",
    "\n",
    "    # remove any previous or obselete antiSMASH annotations to minimise incompatabilities\n",
    "    for record in records:\n",
    "        strip_record(record)\n",
    "\n",
    "    logging.debug(\"Converting records from biopython to secmet\")\n",
    "    try:\n",
    "        records = [Record.from_biopython(record, taxon) for record in records]\n",
    "    except SecmetInvalidInputError as err:\n",
    "        raise AntismashInputError(str(err)) from err\n",
    "\n",
    "    # if parsable by secmet, it has a better context on what to strip, so run\n",
    "    # the secmet stripping to ensure there's no surprises\n",
    "    for record in records:\n",
    "        record.strip_antismash_annotations()\n",
    "\n",
    "    return records\n",
    "\n",
    "\n",
    "def strip_record(record: SeqRecord) -> SeqRecord:\n",
    "    \"\"\" Discard antismash specific features and feature qualifiers\n",
    "\n",
    "        Arguments:\n",
    "            record: the SeqRecord object to clean\n",
    "\n",
    "        Returns\n",
    "            the same SeqRecord instance that was given\n",
    "    \"\"\"\n",
    "    logging.debug(\"Stripping antiSMASH features and annotations from record: %s\", record.id)\n",
    "    old_as_types = {\"cluster\", \"cluster_border\", \"cluster_core\"}\n",
    "    current_as_types = {\"protocluster\", \"proto_core\", \"cand_cluster\", \"subregion\",\n",
    "                        \"region\", \"PFAM_domain\", \"aSDomain\", \"promoter\"}\n",
    "\n",
    "    as_cds_qualifiers = {\n",
    "        \"aSProdPred\",  # <= aS4\n",
    "        \"sec_met\",  # <= aS4, aS5-alpha\n",
    "        \"gene_functions\",  # aS5-beta\n",
    "        \"sec_met_domains\",  # aS5-beta\n",
    "        \"gene_kind\",  # aS5-beta\n",
    "        \"NRPS_PKS\",  # as5.0.0\n",
    "    }\n",
    "\n",
    "    # remove any record-level comment\n",
    "    if \"structured_comment\" in record.annotations:\n",
    "        record.annotations[\"structured_comment\"].pop(\"antiSMASH-Data\", None)\n",
    "        if not record.annotations[\"structured_comment\"]:\n",
    "            record.annotations.pop(\"structured_comment\")\n",
    "\n",
    "    # remove relevant features and qualifiers\n",
    "    kept_features = []\n",
    "    for feature in record.features:\n",
    "        # remove anything explicitly added by antismash, this will catch shared types (e.g. CDS_motif)\n",
    "        if \"antismash\" in [tool.lower() for tool in feature.qualifiers.get(\"tool\", [])]:\n",
    "            continue\n",
    "        if feature.qualifiers.get(\"aSTool\"):\n",
    "            continue\n",
    "        # then check types explicitly as they can be too old to contain the tool\n",
    "        if feature.type in old_as_types:\n",
    "            continue  # remove to avoid confusion\n",
    "        if feature.type in current_as_types:\n",
    "            continue\n",
    "\n",
    "        # old CDS_motif annotations didn't include a \"tool\" qualifier, so clean them up\n",
    "        if feature.type == \"CDS_motif\":\n",
    "            # remove NRPS/PKS domain annotations\n",
    "            notes = [note for note in feature.qualifiers.get(\"note\", []) if not note.startswith(\"NRPS/PKS Motif\")]\n",
    "            # ensure notes isn't left in but empty\n",
    "            if not notes:\n",
    "                feature.qualifiers.pop(\"note\", [])\n",
    "            else:\n",
    "                feature.qualifiers[\"note\"] = notes\n",
    "            # remove any feature that now has no qualifiers\n",
    "            if not feature.qualifiers:\n",
    "                continue\n",
    "\n",
    "        # since CDS features can have a lot of old annotations, clean them up too\n",
    "        if feature.type == \"CDS\":\n",
    "            for qual in as_cds_qualifiers:\n",
    "                feature.qualifiers.pop(qual, None)\n",
    "            notes = feature.qualifiers.get(\"note\")\n",
    "            if notes:  # smcogs\n",
    "                notes = [note for note in notes if not note.startswith(\"smCOG\")]\n",
    "                feature.qualifiers[\"note\"] = notes\n",
    "\n",
    "        kept_features.append(feature)\n",
    "\n",
    "    record.features = kept_features\n",
    "\n",
    "    return record\n",
    "\n",
    "\n",
    "def ensure_cds_info(genefinding: Callable[[Record, Any], None], sequence: Record,\n",
    "                    **kwargs: Dict[str, Any]) -> Record:\n",
    "    \"\"\" Ensures the given record has CDS features with unique locus tags.\n",
    "        CDS features are retrieved from GFF file or via genefinding, depending\n",
    "        on antismash options.\n",
    "\n",
    "        Records without CDS features will have their skip flag marked.\n",
    "\n",
    "        Arguments:\n",
    "            genefinding: the relevant run_on_record(record, options) function to\n",
    "                         use for finding genes if no GFF file being used\n",
    "            record: the Record instance to ensure CDS features for\n",
    "\n",
    "        Returns:\n",
    "            the Record instance provided\n",
    "    \"\"\"\n",
    "    if sequence.skip:\n",
    "        return sequence\n",
    "    options = get_config()\n",
    "    if len(options) == 0:  # inside a parallel function where config doesn't pickle correctly\n",
    "        new = Config(kwargs)\n",
    "        assert isinstance(new, ConfigType)\n",
    "        options = new\n",
    "    if not sequence.get_cds_features():\n",
    "        if not options.genefinding_gff3 and options.genefinding_tool != \"none\":\n",
    "            logging.info(\"No CDS features found in record %r, running gene finding.\", sequence.id)\n",
    "            genefinding(sequence, options)\n",
    "        if not sequence.get_cds_features():\n",
    "            logging.info(\"No genes found, skipping record\")\n",
    "            sequence.skip = \"No genes found\"\n",
    "            return sequence\n",
    "    return sequence\n",
    "\n",
    "\n",
    "def filter_records_by_name(sequences: List[Record], target: str) -> None:\n",
    "    \"\"\" Mark records as skipped if their id does not match the given target or\n",
    "        they are above .\n",
    "\n",
    "        If the target is an empty string, all records will match.\n",
    "        If not records match, an error will be raised.\n",
    "\n",
    "        Arguments:\n",
    "            sequences: the Records to filter\n",
    "            target: the name to match, must be exact\n",
    "\n",
    "        Returns:\n",
    "            None\n",
    "    \"\"\"\n",
    "    if not target:\n",
    "        return\n",
    "\n",
    "    logging.debug(\"Limiting to record id: %s\", target)\n",
    "\n",
    "    # run the filter\n",
    "    matching_filter = 0\n",
    "    for sequence in sequences:\n",
    "        if sequence.id != target:\n",
    "            sequence.skip = \"did not match filter: %s\" % target\n",
    "        else:\n",
    "            matching_filter += 1\n",
    "\n",
    "    if matching_filter == 0:\n",
    "        logging.error(\"No sequences matched filter: %s\", target)\n",
    "        raise AntismashInputError(\"no sequences matched filter: %s\" % target)\n",
    "\n",
    "    logging.info(\"Skipped %d sequences not matching filter: %s\",\n",
    "                 len(sequences) - matching_filter, target)\n",
    "\n",
    "\n",
    "def filter_records_by_count(records: List[Record], maximum: int) -> bool:\n",
    "    \"\"\" Mark all records after the first 'maximum' non-skipped records as skipped.\n",
    "\n",
    "        If maximum is -1, no records will be skipped due to count.\n",
    "\n",
    "    Arguments:\n",
    "        records: the Records to filter\n",
    "        maximum: the maximum number of records to run\n",
    "\n",
    "    Returns:\n",
    "        True if any records were marked as skipped due to hitting the limit\n",
    "    \"\"\"\n",
    "    if maximum == -1:\n",
    "        return False\n",
    "\n",
    "    limit_hit = False\n",
    "    meaningful = 0\n",
    "    for record in records:\n",
    "        if record.skip:  # don't count any records that are already skipped\n",
    "            continue\n",
    "        meaningful += 1\n",
    "        if meaningful > maximum:\n",
    "            limit_hit = True\n",
    "            record.skip = \"skipping all but first {0} meaningful records (--limit {0}) \".format(maximum)\n",
    "\n",
    "    return limit_hit\n",
    "\n",
    "\n",
    "def pre_process_sequences(sequences: List[Record], options: ConfigType, genefinding: AntismashModule) -> List[Record]:\n",
    "    \"\"\" hmm\n",
    "\n",
    "        - gaps removed\n",
    "        - record ids adjusted to be unique\n",
    "        - record ids are valid\n",
    "\n",
    "        Note: Record instances will be altered in-place.\n",
    "\n",
    "        Arguments:\n",
    "            sequences: the secmet.Record instances to process\n",
    "            options: an antismash Config instance\n",
    "            genefinding: the module to use for genefinding, must have\n",
    "                         run_on_record() implemented\n",
    "\n",
    "        Returns:\n",
    "            A list of altered secmet.Record\n",
    "    \"\"\"\n",
    "    logging.debug(\"Preprocessing %d sequences\", len(sequences))\n",
    "\n",
    "    # catch WGS master or supercontig entries\n",
    "    if records_contain_shotgun_scaffolds(sequences):\n",
    "        raise AntismashInputError(\"incomplete whole genome shotgun records are not supported\")\n",
    "\n",
    "    for i, seq in enumerate(sequences):\n",
    "        seq.record_index = i + 1  # 1-indexed\n",
    "\n",
    "    checking_required = not (options.reuse_results or options.skip_sanitisation)\n",
    "\n",
    "    # keep sequences as clean as possible and make sure they're valid\n",
    "    if checking_required:\n",
    "        logging.debug(\"Sanitising record ids and sequences\")\n",
    "        # Ensure all records have unique names\n",
    "        all_record_ids = {seq.id for seq in sequences}\n",
    "        if len(all_record_ids) < len(sequences):\n",
    "            all_record_ids = set()\n",
    "            for record in sequences:\n",
    "                if record.id in all_record_ids:\n",
    "                    record.original_id = record.id\n",
    "                    record.id = generate_unique_id(record.id, all_record_ids)[0]\n",
    "                all_record_ids.add(record.id)\n",
    "            assert len(all_record_ids) == len(sequences), \"%d != %d\" % (len(all_record_ids), len(sequences))\n",
    "        # Ensure all records have valid names\n",
    "        for record in sequences:\n",
    "            fix_record_name_id(record, all_record_ids, options.allow_long_headers)\n",
    "        if len(sequences) == 1:\n",
    "            sequences = [sanitise_sequence(sequences[0])]\n",
    "        else:\n",
    "            sequences = parallel_function(sanitise_sequence, ([record] for record in sequences))\n",
    "\n",
    "    for record in sequences:\n",
    "        if record.skip or not record.seq:\n",
    "            logging.warning(\"Record %s has no sequence, skipping.\", record.id)\n",
    "        if not record.id:\n",
    "            raise AntismashInputError(\"record has no name\")\n",
    "\n",
    "    # skip anything not matching the filter\n",
    "    filter_records_by_name(sequences, options.limit_to_record)\n",
    "\n",
    "    # Now remove small contigs < minimum length again\n",
    "    logging.debug(\"Removing sequences smaller than %d bases\", options.minlength)\n",
    "    for sequence in sequences:\n",
    "        if len(sequence.seq) < options.minlength:\n",
    "            sequence.skip = \"smaller than minimum length (%d)\" % options.minlength\n",
    "\n",
    "    # Make sure we don't waste weeks of runtime on huge records, unless requested by the user\n",
    "    limit_hit = filter_records_by_count(sequences, options.limit)\n",
    "    if limit_hit:\n",
    "        logging.warning(\"Only analysing the first %d records (increase via --limit)\", options.limit)\n",
    "    update_config({\"triggered_limit\": limit_hit})\n",
    "\n",
    "    if checking_required:\n",
    "        # ensure CDS features have all relevant information\n",
    "        logging.debug(\"Ensuring records have CDS features with all required information\")\n",
    "        assert hasattr(genefinding, \"run_on_record\")\n",
    "        genefinding_opts = {key: val for key, val in options if key.startswith(\"genefinding\")}\n",
    "        genefinding_opts[\"taxon\"] = options.taxon\n",
    "        partial = functools.partial(ensure_cds_info, genefinding.run_on_record, **genefinding_opts)\n",
    "        sequences = parallel_function(partial, ([sequence] for sequence in sequences))\n",
    "\n",
    "    if all(sequence.skip for sequence in sequences):\n",
    "        raise AntismashInputError(\"all records skipped\")\n",
    "\n",
    "    return sequences\n",
    "\n",
    "\n",
    "def sanitise_sequence(record: Record) -> Record:\n",
    "    \"\"\" Ensures all sequences use N for gaps instead of -, and that all other\n",
    "        characters are A, C, G, T, or N\n",
    "\n",
    "        Arguments:\n",
    "            records: the secmet.Records to alter\n",
    "\n",
    "        Returns:\n",
    "            the same Record instance as given\n",
    "    \"\"\"\n",
    "    has_real_content = False\n",
    "    sanitised = []\n",
    "    for char in record.seq.upper():\n",
    "        if char == \"-\":\n",
    "            continue\n",
    "        if char in \"ACGT\":\n",
    "            sanitised.append(char)\n",
    "            has_real_content = True\n",
    "        else:\n",
    "            sanitised.append(\"N\")\n",
    "    record.seq = Seq(\"\".join(sanitised))\n",
    "    if not has_real_content:\n",
    "        record.skip = \"contains no sequence\"\n",
    "    return record\n",
    "\n",
    "\n",
    "def trim_sequence(record: SeqRecord, start: int, end: int) -> SeqRecord:\n",
    "    \"\"\" Trims a record to the range given\n",
    "\n",
    "        Arguments:\n",
    "            record: the Bio.SeqRecord to trim\n",
    "            start: the start position (inclusive)\n",
    "            end: the end position (exclusive)\n",
    "\n",
    "        Returns:\n",
    "            A new, shortened Bio.SeqRecord instance\n",
    "    \"\"\"\n",
    "    if start >= len(record):\n",
    "        raise ValueError('Specified analysis start point of %r is outside record' % start)\n",
    "    if end > len(record):\n",
    "        raise ValueError('Specified analysis end point of %r is outside record' % end)\n",
    "    if -1 < end <= start:\n",
    "        raise ValueError(\"Trim region start cannot be greater than or equal to end\")\n",
    "\n",
    "    start = max(start, 0)\n",
    "    if end < 0:\n",
    "        end = len(record)\n",
    "    new = record[start:end]\n",
    "    new.annotations = record.annotations  # preserve old annotations that biopython ignores\n",
    "    return new\n",
    "\n",
    "\n",
    "def records_contain_shotgun_scaffolds(records: List[Record]) -> bool:\n",
    "    \"\"\" Check if given records contain a WGS master record or supercontig record\n",
    "\n",
    "        Arguments:\n",
    "            records: an iterable of secmet.Record\n",
    "\n",
    "        Returns:\n",
    "            True if one of the given records is a WGS or supercontig record\n",
    "    \"\"\"\n",
    "    for record in records:\n",
    "        if isinstance(record.seq, UnknownSeq) and ('wgs_scafld' in record.annotations\n",
    "                                                   or 'wgs' in record.annotations\n",
    "                                                   or 'contig' in record.annotations):\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "\n",
    "def fix_record_name_id(record: Record, all_record_ids: Set[str],\n",
    "                       allow_long_names: bool = False) -> None:\n",
    "    \"\"\" Changes a record's name and id to be no more than 16 characters long,\n",
    "        so it can be used in GenBank files.\n",
    "\n",
    "        If record name is too long, the prefix c000X is used\n",
    "\n",
    "        Arguments:\n",
    "            record: the record to alter\n",
    "            all_record_ids: a set of all known record ids\n",
    "            allow_long_names: whether names longer than 16 characters are allowed\n",
    "\n",
    "        Returns:\n",
    "            None\n",
    "    \"\"\"\n",
    "\n",
    "    def _shorten_ids(idstring: str) -> str:\n",
    "        contigstrmatch = re.search(r\"onti?g?(\\d+)\\b\", idstring)\n",
    "        if not contigstrmatch:\n",
    "            # if there is a substring \"[Ss]caf(fold)XXX\" use this number\n",
    "            contigstrmatch = re.search(r\"caff?o?l?d?(\\d+)\\b\", idstring)\n",
    "        if not contigstrmatch:\n",
    "            # if there is a substring \" cXXX\" use this number\n",
    "            contigstrmatch = re.search(r\"\\bc(\\d+)\\b\", idstring)\n",
    "        if contigstrmatch:\n",
    "            contig_no = int(contigstrmatch.group(1))\n",
    "        else:\n",
    "            # if the contig number cannot be parsed out, just count the contigs from 1 to n\n",
    "            assert isinstance(record.record_index, int)\n",
    "            contig_no = record.record_index\n",
    "\n",
    "        return \"c{ctg:05d}_{origid}..\".format(ctg=contig_no, origid=idstring[:7])\n",
    "\n",
    "    old_id = record.id\n",
    "\n",
    "    if len(record.id) > 16 and not allow_long_names:\n",
    "        # Check if it is a RefSeq accession number like NZ_AMZN01000079.1 that\n",
    "        # is too long just because of the version number behind the dot\n",
    "        if (record.id[-2] == \".\" and\n",
    "                record.id.count(\".\") == 1 and\n",
    "                len(record.id.partition(\".\")[0]) <= 16 and\n",
    "                record.id.partition(\".\")[0] not in all_record_ids):\n",
    "            record.id = record.id.partition(\".\")[0]\n",
    "            all_record_ids.add(record.id)\n",
    "        else:  # Check if the ID suggested by _shorten_ids is unique\n",
    "            if _shorten_ids(old_id) not in all_record_ids:\n",
    "                name = _shorten_ids(old_id)\n",
    "            else:\n",
    "                name, _ = generate_unique_id(record.id[:12], all_record_ids, max_length=16)\n",
    "            record.id = name\n",
    "            all_record_ids.add(name)\n",
    "\n",
    "        logging.warning('Fasta header too long: renamed \"%s\" to \"%s\"', old_id, record.id)\n",
    "\n",
    "    if len(record.name) > 16 and not allow_long_names:\n",
    "        record.name = _shorten_ids(record.name)\n",
    "\n",
    "    if 'accession' in record.annotations and \\\n",
    "       len(record.annotations['accession']) > 16:\n",
    "        acc = record.annotations['accession']\n",
    "\n",
    "        record.annotations['accession'] = _shorten_ids(acc)\n",
    "\n",
    "    # Remove illegal characters from name: otherwise, file cannot be written\n",
    "    illegal_chars = set('''!\"#$%&()*+,:;=>?@[]^`'{|}/ ''')\n",
    "    for char in record.id:\n",
    "        if char in illegal_chars:\n",
    "            record.id = record.id.replace(char, \"\")\n",
    "    for char in record.name:\n",
    "        if char in illegal_chars:\n",
    "            record.name = record.name.replace(char, \"\")\n",
    "\n",
    "    if not record.original_id and old_id != record.id:\n",
    "        record.original_id = old_id\n",
    "\n",
    "\n",
    "def generate_unique_id(prefix: str, existing_ids: Set[str], start: int = 0,\n",
    "                       max_length: int = -1) -> Tuple[str, int]:\n",
    "    \"\"\" Generate a identifier of the form prefix_num, e.g. seq_15.\n",
    "\n",
    "        Does *not* add the generated prefix to current identifiers\n",
    "\n",
    "        Args:\n",
    "            prefix: The text portion of the name.\n",
    "            existing_ids: The current identifiers to avoid collision with.\n",
    "            start: An integer to start counting at (default: 0)\n",
    "            max_length: The maximum length allowed for the identifier,\n",
    "                        values less than 1 are considerd to be no limit.\n",
    "\n",
    "        Returns:\n",
    "            A tuple of the identifier generated and the value of the counter\n",
    "                at the time the identifier was generated, e.g. (\"seq_15\", 15)\n",
    "\n",
    "    \"\"\"\n",
    "    counter = int(start)\n",
    "    existing_ids = set(existing_ids)\n",
    "    max_length = int(max_length)\n",
    "\n",
    "    name = f\"{prefix}_{counter}\"\n",
    "    while name in existing_ids:\n",
    "        counter += 1\n",
    "        name = f\"{prefix}_{counter}\"\n",
    "    if 0 < max_length < len(name):\n",
    "        raise RuntimeError(\"Could not generate unique id for %s after %d iterations\" % (prefix, counter - start))\n",
    "    return name, counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cc6a9f34-e3dc-435b-8e3a-36c9dd9ec1f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# License: GNU Affero General Public License v3 or later\n",
    "# A copy of GNU AGPL v3 should have been included in this software package in LICENSE.txt.\n",
    "\n",
    "\"\"\" Reads GFF files and updates records with the contained information.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "import logging\n",
    "from typing import Dict, IO, List, Set\n",
    "\n",
    "from Bio.SeqFeature import FeatureLocation, CompoundLocation, SeqFeature\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from BCBio import GFF\n",
    "\n",
    "from antismash.common.errors import AntismashInputError\n",
    "\n",
    "# whether to use phase (codon start) to modify reported locations\n",
    "# Augustus, NCBI, and glimmerhmm report phase but have already adjusted the\n",
    "# locations and since they're the bulk of inputs, disable further modification\n",
    "MODIFY_LOCATIONS_BY_PHASE = False\n",
    "\n",
    "\n",
    "def check_gff_suitability(gff_file: str, sequences: List[SeqRecord]) -> None:\n",
    "    \"\"\"\n",
    "        Checks that the provided GFF3 file is acceptable\n",
    "\n",
    "        If only a single record is contained in both sequences and GFF, they\n",
    "        are assumed to be the same.\n",
    "\n",
    "        Arguments:\n",
    "            gff_file: the path of the GFF file to check\n",
    "            sequences: a list of SeqRecords\n",
    "\n",
    "        Returns:\n",
    "            None\n",
    "    \"\"\"\n",
    "    try:\n",
    "        examiner = GFF.GFFExaminer()\n",
    "        # file handle is automatically closed by GFF lib\n",
    "        # pylint:disable=consider-using-with\n",
    "        gff_data = examiner.available_limits(open(gff_file))\n",
    "        # Check if at least one GFF locus appears in sequence\n",
    "        gff_ids = set(n[0] for n in gff_data['gff_id'])\n",
    "\n",
    "        if len(gff_ids) == 1 and len(sequences) == 1:\n",
    "            # If both inputs only have one record, assume is the same,\n",
    "            # but first check coordinate compatibility\n",
    "            logging.info(\"GFF3 and sequence have only one record. Assuming is \"\n",
    "                         \"the same as long as coordinates are compatible.\")\n",
    "            limit_info = dict(gff_type=['CDS'])\n",
    "\n",
    "            record_iter = GFF.parse(open(gff_file), limit_info=limit_info)\n",
    "            try:\n",
    "                record = next(record_iter)\n",
    "            except StopIteration:\n",
    "                raise AntismashInputError(\"could not parse records from GFF3 file\")\n",
    "\n",
    "            if not record.features:\n",
    "                raise AntismashInputError('GFF3 record %s contains no features' % record.id)\n",
    "\n",
    "            coord_max = max([n.location.end.real for n in record.features])\n",
    "            if coord_max > len(sequences[0]):\n",
    "                logging.error('GFF3 record and sequence coordinates are not compatible.')\n",
    "                raise AntismashInputError('incompatible GFF record and sequence coordinates')\n",
    "\n",
    "        elif not gff_ids.intersection({seq.id for seq in sequences}):\n",
    "            logging.error('No GFF3 record IDs match any sequence record IDs.')\n",
    "            raise AntismashInputError(\"GFF3 record IDs don't match sequence file record IDs.\")\n",
    "\n",
    "        # Check GFF contains CDSs\n",
    "        if not ('CDS',) in gff_data['gff_type']:\n",
    "            logging.error('GFF3 does not contain any CDS.')\n",
    "            raise AntismashInputError(\"no CDS features in GFF3 file.\")\n",
    "\n",
    "        # Check CDS are childless but not parentless\n",
    "        if 'CDS' in set(n for key in examiner.parent_child_map(open(gff_file)) for n in key):\n",
    "            logging.error('GFF3 structure is not suitable. CDS features must be childless but not parentless.')\n",
    "            raise AntismashInputError('GFF3 structure is not suitable.')\n",
    "        # pylint:enable=consider-using-with\n",
    "\n",
    "    except AssertionError as err:\n",
    "        # usually the assertion \"assert len(parts) >= 8, line\"\n",
    "        # so strip the newline and improve the error message\n",
    "        message = str(err).strip()\n",
    "        raise AntismashInputError(\"parsing GFF failed with invalid format: %r\" % message) from err\n",
    "\n",
    "\n",
    "def get_features_from_file(handle: IO) -> Dict[str, List[SeqFeature]]:\n",
    "    \"\"\" Generates new SeqFeatures from a GFF file.\n",
    "\n",
    "        Arguments:\n",
    "            handle: a file handle/stream with the GFF contents\n",
    "\n",
    "        Returns:\n",
    "            a dictionary mapping record ID to a list of SeqFeatures for that record\n",
    "    \"\"\"\n",
    "    try:\n",
    "        gff_records = list(GFF.parse(handle))\n",
    "    except Exception as err:\n",
    "        raise AntismashInputError(\"could not parse records from GFF3 file\") from err\n",
    "\n",
    "    results = {}\n",
    "    for gff_record in gff_records:\n",
    "        features = []\n",
    "        for feature in gff_record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                new_features = [feature]\n",
    "            else:\n",
    "                new_features = check_sub(feature)\n",
    "                if feature.type == \"gene\":\n",
    "                    features.append(feature)\n",
    "                if not new_features:\n",
    "                    continue\n",
    "\n",
    "            name = feature.id\n",
    "            locus_tag = feature.qualifiers.get(\"locus_tag\")\n",
    "\n",
    "            for qtype in [\"gene\", \"name\", \"Name\"]:\n",
    "                if qtype in feature.qualifiers:\n",
    "                    name_tmp = feature.qualifiers[qtype][0]\n",
    "                    # Assume name/Name to be sane if they don't contain a space\n",
    "                    if \" \" in name_tmp:\n",
    "                        continue\n",
    "                    name = name_tmp\n",
    "                    break\n",
    "\n",
    "            multiple_cds = len(list(filter(lambda x: x.type == \"CDS\", new_features))) > 1\n",
    "            for i, new_feature in enumerate(new_features):\n",
    "                variant = name\n",
    "                if new_feature.type == \"CDS\" and multiple_cds:\n",
    "                    variant = \"{0}_{1}\".format(name, i)\n",
    "                new_feature.qualifiers['gene'] = [variant]\n",
    "                if locus_tag is not None:\n",
    "                    new_feature.qualifiers[\"locus_tag\"] = locus_tag\n",
    "                features.append(new_feature)\n",
    "        results[gff_record.id] = features\n",
    "    return results\n",
    "\n",
    "\n",
    "def run(gff_file: str) -> Dict[str, List[SeqFeature]]:\n",
    "    \"\"\" The entry point of gff_parser.\n",
    "        Generates new features and adds them to the provided record.\n",
    "\n",
    "        Arguments:\n",
    "            options: an antismash.Config object, used only for fetching the GFF path\n",
    "\n",
    "        Returns:\n",
    "            a dictionary mapping record ID to a list of SeqFeatures in that record\n",
    "    \"\"\"\n",
    "    with open(gff_file) as handle:\n",
    "        return get_features_from_file(handle)\n",
    "\n",
    "\n",
    "def generate_details_from_subfeature(sub_feature: SeqFeature,\n",
    "                                     existing_qualifiers: Dict[str, List[str]],\n",
    "                                     locations: List[FeatureLocation],\n",
    "                                     trans_locations: List[FeatureLocation]) -> Set[str]:\n",
    "    \"\"\" Finds the locations of a subfeature and any mismatching qualifiers\n",
    "\n",
    "        Arguments:\n",
    "            sub_feature: the GFF subfeature to work on\n",
    "            existing_qualifiers: a dict of any existing qualifiers from other\n",
    "                                 subfeatures\n",
    "            locations: a list of any existing FeatureLocations from other\n",
    "                       subfeatures\n",
    "            trans_locations: a list of any existing FeatureLocations for\n",
    "                             translations\n",
    "\n",
    "        Returns:\n",
    "            a set of qualifiers from the subfeature for which an existing\n",
    "            qualifier existed but had a different value\n",
    "    \"\"\"\n",
    "    mismatching_qualifiers = set()\n",
    "    start = sub_feature.location.start.real\n",
    "    end = sub_feature.location.end.real\n",
    "    if MODIFY_LOCATIONS_BY_PHASE:\n",
    "        phase = int(sub_feature.qualifiers.get('phase', [0])[0])\n",
    "        if sub_feature.strand == 1:\n",
    "            start += phase\n",
    "        else:\n",
    "            end -= phase\n",
    "    try:\n",
    "        locations.append(FeatureLocation(start, end, strand=sub_feature.strand))\n",
    "    except ValueError as err:\n",
    "        raise AntismashInputError(str(err)) from err\n",
    "    # Make sure CDSs lengths are multiple of three. Otherwise extend to next full codon.\n",
    "    # This only applies for translation.\n",
    "    modulus = (end - start) % 3\n",
    "    if modulus and sub_feature.strand == 1:\n",
    "        end += 3 - modulus\n",
    "    elif modulus and sub_feature.strand == -1:\n",
    "        start -= 3 - modulus\n",
    "    trans_locations.append(FeatureLocation(start, end, strand=sub_feature.strand))\n",
    "    # For split features (CDSs), the final feature will have the same qualifiers as the children ONLY if\n",
    "    # they're the same, i.e.: all children have the same \"protein_ID\" (key and value).\n",
    "    for qual in sub_feature.qualifiers:\n",
    "        if qual not in existing_qualifiers:\n",
    "            existing_qualifiers[qual] = sub_feature.qualifiers[qual]\n",
    "        elif existing_qualifiers[qual] != sub_feature.qualifiers[qual]:\n",
    "            mismatching_qualifiers.add(qual)\n",
    "    return mismatching_qualifiers\n",
    "\n",
    "\n",
    "def check_sub(feature: SeqFeature) -> List[SeqFeature]:\n",
    "    \"\"\" Recursively checks a GFF feature for any subfeatures and generates any\n",
    "        appropriate SeqFeature instances from them.\n",
    "    \"\"\"\n",
    "    new_features = []\n",
    "    locations: List[FeatureLocation] = []\n",
    "    trans_locations: List[FeatureLocation] = []\n",
    "    qualifiers: Dict[str, List[str]] = {}\n",
    "    mismatching_qualifiers: Set[str] = set()\n",
    "    for sub in feature.sub_features:\n",
    "        if sub.type != \"CDS\":\n",
    "            new_features.append(sub)\n",
    "        if sub.sub_features:  # If there are sub_features, go deeper\n",
    "            new_features.extend(check_sub(sub))\n",
    "        elif sub.type == 'CDS':\n",
    "            sub_mismatch = generate_details_from_subfeature(sub, qualifiers,\n",
    "                                                            locations, trans_locations)\n",
    "            mismatching_qualifiers.update(sub_mismatch)\n",
    "\n",
    "    for qualifier in mismatching_qualifiers:\n",
    "        del qualifiers[qualifier]\n",
    "    if 'Parent' in qualifiers:\n",
    "        del qualifiers['Parent']\n",
    "\n",
    "    # if nothing to work on\n",
    "    if not new_features and not locations:\n",
    "        return []\n",
    "\n",
    "    # Only works in tip of the tree, when there's no new_feature built yet. If there is,\n",
    "    # it means the script just came out of a check_sub and it's ready to return.\n",
    "    if locations:\n",
    "        new_loc = locations[0]\n",
    "        # construct a compound location if required\n",
    "        if len(locations) > 1:\n",
    "            locations = sorted(locations, key=lambda x: x.start.real)\n",
    "            trans_locations = sorted(trans_locations, key=lambda x: x.start.real)\n",
    "            if locations[0].strand == 1:\n",
    "                new_loc = CompoundLocation(locations)\n",
    "            else:\n",
    "                new_loc = CompoundLocation(list(reversed(locations)))\n",
    "                trans_locations = list(reversed(trans_locations))\n",
    "        new_feature = SeqFeature(new_loc)\n",
    "        new_feature.qualifiers = qualifiers\n",
    "        new_feature.type = 'CDS'\n",
    "        new_features.append(new_feature)\n",
    "\n",
    "    return new_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f45c6c2d-c2f0-4946-8c8b-f9aca876ebeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "records = records_prparse_input_sequence(\"../data/interim/fasta/1097422.fna\",\n",
    "                                gff_file=\"../data/interim/prokka/1097422/1097422.gff\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7b4dbde7-d351-4a09-ad26-ee3a087ce7c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[mRNA([51398:53084](-)),\n",
       " exon([52633:53084](-)),\n",
       " five_prime_UTR([52958:53084](-)),\n",
       " exon([52396:52575](-)),\n",
       " exon([52168:52312](-)),\n",
       " exon([51398:52117](-)),\n",
       " three_prime_UTR([51398:51511](-)),\n",
       " mRNA([53217:55265](+)),\n",
       " exon([53217:53553](+)),\n",
       " five_prime_UTR([53217:53431](+)),\n",
       " exon([53624:53935](+)),\n",
       " exon([53988:54130](+)),\n",
       " exon([54181:55265](+)),\n",
       " three_prime_UTR([54824:55265](+)),\n",
       " mRNA([64047:65730](-)),\n",
       " exon([65177:65730](-)),\n",
       " five_prime_UTR([65690:65730](-)),\n",
       " exon([64677:65126](-)),\n",
       " exon([64047:64622](-)),\n",
       " three_prime_UTR([64047:64309](-)),\n",
       " mRNA([69003:69915](+)),\n",
       " exon([69003:69410](+)),\n",
       " five_prime_UTR([69003:69169](+)),\n",
       " exon([69468:69915](+)),\n",
       " three_prime_UTR([69842:69915](+)),\n",
       " mRNA([78417:79670](-)),\n",
       " exon([78831:79670](-)),\n",
       " five_prime_UTR([79486:79670](-)),\n",
       " exon([78417:78777](-)),\n",
       " three_prime_UTR([78417:78694](-)),\n",
       " mRNA([88824:90861](-)),\n",
       " exon([90532:90861](-)),\n",
       " five_prime_UTR([90678:90861](-)),\n",
       " exon([88824:90474](-)),\n",
       " three_prime_UTR([88824:88850](-)),\n",
       " mRNA([128022:130081](-)),\n",
       " exon([128022:130081](-)),\n",
       " five_prime_UTR([129891:130081](-)),\n",
       " three_prime_UTR([128022:128325](-)),\n",
       " mRNA([140319:143285](-)),\n",
       " exon([142904:143285](-)),\n",
       " five_prime_UTR([143192:143285](-)),\n",
       " exon([140319:142850](-)),\n",
       " three_prime_UTR([140319:140354](-)),\n",
       " mRNA([185348:187205](-)),\n",
       " exon([186740:187205](-)),\n",
       " five_prime_UTR([186901:187205](-)),\n",
       " exon([186419:186654](-)),\n",
       " exon([185348:186348](-)),\n",
       " three_prime_UTR([185348:185565](-)),\n",
       " mRNA([189149:191191](-)),\n",
       " exon([190974:191191](-)),\n",
       " five_prime_UTR([191097:191191](-)),\n",
       " exon([189149:190906](-)),\n",
       " three_prime_UTR([189149:189433](-)),\n",
       " mRNA([212039:213530](+)),\n",
       " exon([212039:213530](+)),\n",
       " five_prime_UTR([212039:212206](+)),\n",
       " three_prime_UTR([213319:213530](+)),\n",
       " mRNA([301387:303020](-)),\n",
       " exon([302652:303020](-)),\n",
       " five_prime_UTR([302767:303020](-)),\n",
       " exon([302202:302576](-)),\n",
       " exon([301732:302152](-)),\n",
       " exon([301387:301669](-)),\n",
       " three_prime_UTR([301387:301576](-)),\n",
       " mRNA([353202:355566](+)),\n",
       " exon([353202:353296](+)),\n",
       " five_prime_UTR([353202:353296](+)),\n",
       " exon([353433:354051](+)),\n",
       " five_prime_UTR([353433:353498](+)),\n",
       " exon([354104:355566](+)),\n",
       " three_prime_UTR([355555:355566](+)),\n",
       " mRNA([366869:368945](-)),\n",
       " exon([368584:368945](-)),\n",
       " five_prime_UTR([368769:368945](-)),\n",
       " exon([368431:368518](-)),\n",
       " exon([368040:368372](-)),\n",
       " exon([367635:367991](-)),\n",
       " exon([366869:367575](-)),\n",
       " three_prime_UTR([366869:367149](-)),\n",
       " mRNA([385378:386638](-)),\n",
       " exon([386620:386638](-)),\n",
       " five_prime_UTR([386620:386638](-)),\n",
       " exon([386101:386557](-)),\n",
       " five_prime_UTR([386351:386557](-)),\n",
       " exon([385378:386015](-)),\n",
       " three_prime_UTR([385378:385407](-)),\n",
       " mRNA([430089:431500](+)),\n",
       " exon([430089:430351](+)),\n",
       " five_prime_UTR([430089:430154](+)),\n",
       " exon([430492:431126](+)),\n",
       " exon([431180:431270](+)),\n",
       " exon([431328:431500](+)),\n",
       " three_prime_UTR([431412:431500](+)),\n",
       " mRNA([443407:445377](-)),\n",
       " exon([444674:445377](-)),\n",
       " five_prime_UTR([444897:445377](-)),\n",
       " exon([443407:444606](-)),\n",
       " three_prime_UTR([443407:443770](-)),\n",
       " mRNA([448807:450702](-)),\n",
       " exon([450261:450702](-)),\n",
       " five_prime_UTR([450675:450702](-)),\n",
       " exon([449995:450207](-)),\n",
       " exon([448807:449930](-)),\n",
       " three_prime_UTR([448807:448894](-)),\n",
       " mRNA([504002:506104](-)),\n",
       " exon([505728:506104](-)),\n",
       " five_prime_UTR([505912:506104](-)),\n",
       " exon([504002:505670](-)),\n",
       " three_prime_UTR([504002:504252](-)),\n",
       " mRNA([532598:537378](-)),\n",
       " exon([537150:537378](-)),\n",
       " five_prime_UTR([537153:537378](-)),\n",
       " exon([536836:537073](-)),\n",
       " exon([536723:536769](-)),\n",
       " exon([536177:536660](-)),\n",
       " exon([534421:536119](-)),\n",
       " exon([534277:534369](-)),\n",
       " exon([533982:534230](-)),\n",
       " exon([533400:533927](-)),\n",
       " exon([532598:533346](-)),\n",
       " three_prime_UTR([532598:532924](-)),\n",
       " mRNA([638134:640812](+)),\n",
       " exon([638134:638517](+)),\n",
       " five_prime_UTR([638134:638232](+)),\n",
       " exon([638565:638775](+)),\n",
       " exon([638827:639339](+)),\n",
       " exon([639390:639618](+)),\n",
       " exon([639667:640812](+)),\n",
       " three_prime_UTR([640475:640812](+)),\n",
       " mRNA([683480:686721](-)),\n",
       " exon([686244:686721](-)),\n",
       " five_prime_UTR([686495:686721](-)),\n",
       " exon([683480:686169](-)),\n",
       " three_prime_UTR([683480:683522](-)),\n",
       " mRNA([695609:697324](+)),\n",
       " exon([695609:695763](+)),\n",
       " five_prime_UTR([695609:695684](+)),\n",
       " exon([695894:696042](+)),\n",
       " exon([696141:697324](+)),\n",
       " three_prime_UTR([696928:697324](+)),\n",
       " mRNA([788095:789242](+)),\n",
       " exon([788095:788204](+)),\n",
       " five_prime_UTR([788095:788163](+)),\n",
       " exon([788251:788378](+)),\n",
       " exon([788427:789242](+)),\n",
       " three_prime_UTR([788979:789242](+)),\n",
       " mRNA([800921:802077](+)),\n",
       " exon([800921:801299](+)),\n",
       " five_prime_UTR([800921:801058](+)),\n",
       " exon([801345:801474](+)),\n",
       " exon([801525:801600](+)),\n",
       " exon([801656:802077](+)),\n",
       " three_prime_UTR([802009:802077](+)),\n",
       " mRNA([881089:882373](-)),\n",
       " exon([881949:882373](-)),\n",
       " five_prime_UTR([882035:882373](-)),\n",
       " exon([881089:881882](-)),\n",
       " three_prime_UTR([881089:881341](-)),\n",
       " mRNA([892263:894263](+)),\n",
       " exon([892263:892544](+)),\n",
       " five_prime_UTR([892263:892516](+)),\n",
       " exon([892616:892691](+)),\n",
       " exon([892748:892825](+)),\n",
       " exon([892878:893037](+)),\n",
       " exon([893085:893205](+)),\n",
       " exon([893261:894106](+)),\n",
       " exon([894158:894263](+)),\n",
       " three_prime_UTR([894255:894263](+)),\n",
       " mRNA([934189:936485](+)),\n",
       " exon([934189:934366](+)),\n",
       " five_prime_UTR([934189:934268](+)),\n",
       " exon([934418:935166](+)),\n",
       " exon([935533:936485](+)),\n",
       " three_prime_UTR([936379:936485](+)),\n",
       " mRNA([957652:958638](-)),\n",
       " exon([958405:958638](-)),\n",
       " five_prime_UTR([958628:958638](-)),\n",
       " exon([957652:958270](-)),\n",
       " three_prime_UTR([957652:957836](-)),\n",
       " mRNA([1060843:1062589](+)),\n",
       " exon([1060843:1060953](+)),\n",
       " five_prime_UTR([1060843:1060894](+)),\n",
       " exon([1061003:1062205](+)),\n",
       " exon([1062279:1062349](+)),\n",
       " exon([1062403:1062589](+)),\n",
       " three_prime_UTR([1062554:1062589](+)),\n",
       " mRNA([1071128:1073951](-)),\n",
       " exon([1073353:1073951](-)),\n",
       " five_prime_UTR([1073635:1073951](-)),\n",
       " exon([1072432:1073277](-)),\n",
       " exon([1071128:1072385](-)),\n",
       " three_prime_UTR([1071128:1071211](-)),\n",
       " mRNA([1150872:1152094](-)),\n",
       " exon([1151951:1152094](-)),\n",
       " five_prime_UTR([1151975:1152094](-)),\n",
       " exon([1151705:1151890](-)),\n",
       " exon([1151227:1151646](-)),\n",
       " exon([1150872:1151159](-)),\n",
       " three_prime_UTR([1150872:1151019](-)),\n",
       " mRNA([1163957:1165208](-)),\n",
       " exon([1165119:1165208](-)),\n",
       " five_prime_UTR([1165132:1165208](-)),\n",
       " exon([1164827:1165019](-)),\n",
       " exon([1164328:1164760](-)),\n",
       " exon([1163957:1164276](-)),\n",
       " three_prime_UTR([1163957:1164262](-)),\n",
       " mRNA([1176323:1177300](-)),\n",
       " exon([1177019:1177300](-)),\n",
       " five_prime_UTR([1177290:1177300](-)),\n",
       " exon([1176844:1176958](-)),\n",
       " exon([1176323:1176782](-)),\n",
       " three_prime_UTR([1176323:1176408](-)),\n",
       " mRNA([1194540:1195996](-)),\n",
       " exon([1195294:1195996](-)),\n",
       " five_prime_UTR([1195729:1195996](-)),\n",
       " exon([1194540:1195228](-)),\n",
       " three_prime_UTR([1194540:1194604](-)),\n",
       " mRNA([1332350:1333115](+)),\n",
       " exon([1332350:1332607](+)),\n",
       " five_prime_UTR([1332350:1332457](+)),\n",
       " exon([1332697:1333115](+)),\n",
       " three_prime_UTR([1332952:1333115](+)),\n",
       " mRNA([1351772:1352760](-)),\n",
       " exon([1351772:1352760](-)),\n",
       " five_prime_UTR([1352658:1352760](-)),\n",
       " three_prime_UTR([1351772:1352001](-)),\n",
       " mRNA([1385709:1386928](-)),\n",
       " exon([1386602:1386928](-)),\n",
       " five_prime_UTR([1386773:1386928](-)),\n",
       " exon([1385907:1386546](-)),\n",
       " exon([1385709:1385841](-)),\n",
       " three_prime_UTR([1385709:1385793](-)),\n",
       " mRNA([1470559:1471359](-)),\n",
       " exon([1470559:1471359](-)),\n",
       " five_prime_UTR([1471289:1471359](-)),\n",
       " three_prime_UTR([1470559:1470833](-)),\n",
       " mRNA([1483525:1484416](-)),\n",
       " exon([1483525:1484416](-)),\n",
       " five_prime_UTR([1484345:1484416](-)),\n",
       " three_prime_UTR([1483525:1483571](-)),\n",
       " mRNA([1574502:1576484](-)),\n",
       " exon([1576260:1576484](-)),\n",
       " five_prime_UTR([1576336:1576484](-)),\n",
       " exon([1576017:1576191](-)),\n",
       " exon([1574502:1575962](-)),\n",
       " three_prime_UTR([1574502:1574625](-)),\n",
       " mRNA([1647873:1650754](+)),\n",
       " exon([1647873:1648384](+)),\n",
       " five_prime_UTR([1647873:1648011](+)),\n",
       " exon([1648442:1650754](+)),\n",
       " three_prime_UTR([1650286:1650754](+)),\n",
       " mRNA([1650626:1651954](-)),\n",
       " exon([1651695:1651954](-)),\n",
       " five_prime_UTR([1651762:1651954](-)),\n",
       " exon([1651508:1651637](-)),\n",
       " exon([1651386:1651452](-)),\n",
       " exon([1650626:1651335](-)),\n",
       " three_prime_UTR([1650626:1650967](-)),\n",
       " mRNA([1741648:1743055](-)),\n",
       " exon([1742954:1743055](-)),\n",
       " five_prime_UTR([1742963:1743055](-)),\n",
       " exon([1742790:1742898](-)),\n",
       " exon([1742145:1742734](-)),\n",
       " exon([1741648:1742029](-)),\n",
       " three_prime_UTR([1741648:1741655](-)),\n",
       " mRNA([1751180:1754553](+)),\n",
       " exon([1751180:1751610](+)),\n",
       " five_prime_UTR([1751180:1751506](+)),\n",
       " exon([1751670:1753633](+)),\n",
       " exon([1753739:1754553](+)),\n",
       " three_prime_UTR([1754075:1754553](+)),\n",
       " mRNA([1756969:1757954](-)),\n",
       " exon([1757468:1757954](-)),\n",
       " five_prime_UTR([1757843:1757954](-)),\n",
       " exon([1756969:1757366](-)),\n",
       " three_prime_UTR([1756969:1757090](-)),\n",
       " mRNA([1770665:1771775](+)),\n",
       " exon([1770665:1770764](+)),\n",
       " five_prime_UTR([1770665:1770764](+)),\n",
       " exon([1770835:1771160](+)),\n",
       " five_prime_UTR([1770835:1770839](+)),\n",
       " exon([1771245:1771640](+)),\n",
       " exon([1771693:1771775](+)),\n",
       " three_prime_UTR([1771742:1771775](+)),\n",
       " mRNA([1791247:1793615](-)),\n",
       " exon([1793471:1793615](-)),\n",
       " five_prime_UTR([1793471:1793615](-)),\n",
       " exon([1792716:1793307](-)),\n",
       " five_prime_UTR([1793125:1793307](-)),\n",
       " exon([1792522:1792662](-)),\n",
       " exon([1791247:1792460](-)),\n",
       " three_prime_UTR([1791247:1791677](-)),\n",
       " mRNA([1799028:1800357](+)),\n",
       " exon([1799028:1799078](+)),\n",
       " five_prime_UTR([1799028:1799078](+)),\n",
       " exon([1799144:1800357](+)),\n",
       " five_prime_UTR([1799144:1799249](+)),\n",
       " three_prime_UTR([1800278:1800357](+)),\n",
       " mRNA([1832712:1835203](-)),\n",
       " exon([1834633:1835203](-)),\n",
       " five_prime_UTR([1834974:1835203](-)),\n",
       " exon([1834381:1834581](-)),\n",
       " exon([1832712:1834334](-)),\n",
       " three_prime_UTR([1832712:1833063](-)),\n",
       " mRNA([1865686:1866459](+)),\n",
       " exon([1865686:1865873](+)),\n",
       " five_prime_UTR([1865686:1865873](+)),\n",
       " exon([1865932:1866459](+)),\n",
       " five_prime_UTR([1865932:1865960](+)),\n",
       " three_prime_UTR([1866401:1866459](+)),\n",
       " mRNA([1918750:1920725](+)),\n",
       " exon([1918750:1919232](+)),\n",
       " five_prime_UTR([1918750:1918897](+)),\n",
       " exon([1919293:1920725](+)),\n",
       " three_prime_UTR([1920659:1920725](+)),\n",
       " mRNA([1935068:1935885](+)),\n",
       " exon([1935068:1935184](+)),\n",
       " five_prime_UTR([1935068:1935148](+)),\n",
       " exon([1935260:1935885](+)),\n",
       " three_prime_UTR([1935800:1935885](+)),\n",
       " mRNA([1946434:1949297](-)),\n",
       " exon([1948905:1949297](-)),\n",
       " five_prime_UTR([1948945:1949297](-)),\n",
       " exon([1948769:1948819](-)),\n",
       " exon([1948641:1948676](-)),\n",
       " exon([1948490:1948566](-)),\n",
       " exon([1948339:1948438](-)),\n",
       " exon([1948165:1948286](-)),\n",
       " exon([1947495:1948107](-)),\n",
       " exon([1947226:1947442](-)),\n",
       " exon([1946434:1947177](-)),\n",
       " three_prime_UTR([1946434:1946626](-)),\n",
       " mRNA([2020947:2021973](-)),\n",
       " exon([2021652:2021973](-)),\n",
       " five_prime_UTR([2021836:2021973](-)),\n",
       " exon([2020947:2021528](-)),\n",
       " three_prime_UTR([2020947:2021016](-)),\n",
       " mRNA([2022318:2024065](-)),\n",
       " exon([2023345:2024065](-)),\n",
       " five_prime_UTR([2023843:2024065](-)),\n",
       " exon([2022412:2023289](-)),\n",
       " exon([2022318:2022352](-)),\n",
       " three_prime_UTR([2022318:2022329](-)),\n",
       " mRNA([2046410:2047244](+)),\n",
       " exon([2046410:2046633](+)),\n",
       " five_prime_UTR([2046410:2046475](+)),\n",
       " exon([2046712:2046818](+)),\n",
       " exon([2046910:2047244](+)),\n",
       " three_prime_UTR([2046994:2047244](+)),\n",
       " mRNA([2054286:2055772](+)),\n",
       " exon([2054286:2054617](+)),\n",
       " five_prime_UTR([2054286:2054354](+)),\n",
       " exon([2054676:2054967](+)),\n",
       " exon([2055032:2055772](+)),\n",
       " three_prime_UTR([2055543:2055772](+)),\n",
       " mRNA([2199575:2200801](+)),\n",
       " exon([2199575:2199689](+)),\n",
       " five_prime_UTR([2199575:2199642](+)),\n",
       " exon([2199742:2200346](+)),\n",
       " exon([2200401:2200801](+)),\n",
       " three_prime_UTR([2200755:2200801](+)),\n",
       " mRNA([2220611:2222899](+)),\n",
       " exon([2220611:2221115](+)),\n",
       " five_prime_UTR([2220611:2220813](+)),\n",
       " exon([2221170:2221294](+)),\n",
       " exon([2221345:2221422](+)),\n",
       " exon([2221472:2222899](+)),\n",
       " three_prime_UTR([2222595:2222899](+)),\n",
       " mRNA([2224337:2226408](+)),\n",
       " exon([2224337:2224812](+)),\n",
       " five_prime_UTR([2224337:2224689](+)),\n",
       " exon([2224876:2225602](+)),\n",
       " exon([2225676:2226408](+)),\n",
       " three_prime_UTR([2226147:2226408](+)),\n",
       " mRNA([2285658:2287494](-)),\n",
       " exon([2287031:2287494](-)),\n",
       " five_prime_UTR([2287406:2287494](-)),\n",
       " exon([2285658:2286969](-)),\n",
       " three_prime_UTR([2285658:2285670](-)),\n",
       " mRNA([2288993:2290308](-)),\n",
       " exon([2290276:2290308](-)),\n",
       " five_prime_UTR([2290286:2290308](-)),\n",
       " exon([2289971:2290227](-)),\n",
       " exon([2288993:2289912](-)),\n",
       " three_prime_UTR([2288993:2289104](-)),\n",
       " mRNA([2327151:2328866](-)),\n",
       " exon([2328767:2328866](-)),\n",
       " five_prime_UTR([2328789:2328866](-)),\n",
       " exon([2327151:2328704](-)),\n",
       " three_prime_UTR([2327151:2327325](-)),\n",
       " mRNA([2376002:2377712](-)),\n",
       " exon([2376370:2377712](-)),\n",
       " five_prime_UTR([2377597:2377712](-)),\n",
       " exon([2376002:2376320](-)),\n",
       " three_prime_UTR([2376002:2376260](-)),\n",
       " mRNA([2388798:2390812](+)),\n",
       " exon([2388798:2389545](+)),\n",
       " five_prime_UTR([2388798:2388879](+)),\n",
       " exon([2389613:2390334](+)),\n",
       " exon([2390387:2390812](+)),\n",
       " three_prime_UTR([2390425:2390812](+)),\n",
       " mRNA([2474135:2475367](+)),\n",
       " exon([2474135:2474169](+)),\n",
       " five_prime_UTR([2474135:2474136](+)),\n",
       " exon([2474223:2474632](+)),\n",
       " exon([2474694:2475367](+)),\n",
       " three_prime_UTR([2475299:2475367](+)),\n",
       " mRNA([2583994:2585952](+)),\n",
       " exon([2583994:2584463](+)),\n",
       " five_prime_UTR([2583994:2584151](+)),\n",
       " exon([2584522:2585109](+)),\n",
       " exon([2585155:2585952](+)),\n",
       " three_prime_UTR([2585618:2585952](+)),\n",
       " mRNA([2601275:2602826](-)),\n",
       " exon([2602514:2602826](-)),\n",
       " five_prime_UTR([2602619:2602826](-)),\n",
       " exon([2602203:2602447](-)),\n",
       " exon([2601992:2602144](-)),\n",
       " exon([2601275:2601942](-)),\n",
       " three_prime_UTR([2601275:2601618](-)),\n",
       " mRNA([2655834:2656978](-)),\n",
       " exon([2656850:2656978](-)),\n",
       " five_prime_UTR([2656850:2656978](-)),\n",
       " exon([2656700:2656747](-)),\n",
       " five_prime_UTR([2656706:2656747](-)),\n",
       " exon([2656622:2656633](-)),\n",
       " exon([2656265:2656526](-)),\n",
       " exon([2656157:2656206](-)),\n",
       " exon([2655834:2656098](-)),\n",
       " three_prime_UTR([2655834:2655960](-)),\n",
       " mRNA([2685565:2686686](+)),\n",
       " exon([2685565:2685934](+)),\n",
       " five_prime_UTR([2685565:2685666](+)),\n",
       " exon([2686076:2686165](+)),\n",
       " exon([2686238:2686389](+)),\n",
       " exon([2686451:2686686](+)),\n",
       " three_prime_UTR([2686633:2686686](+)),\n",
       " mRNA([2709098:2711012](+)),\n",
       " exon([2709098:2709253](+)),\n",
       " five_prime_UTR([2709098:2709133](+)),\n",
       " exon([2709315:2709393](+)),\n",
       " exon([2709466:2710309](+)),\n",
       " exon([2710379:2710509](+)),\n",
       " exon([2710572:2711012](+)),\n",
       " three_prime_UTR([2710901:2711012](+)),\n",
       " mRNA([2711477:2713085](-)),\n",
       " exon([2712824:2713085](-)),\n",
       " five_prime_UTR([2712852:2713085](-)),\n",
       " exon([2712694:2712752](-)),\n",
       " exon([2712457:2712629](-)),\n",
       " exon([2711477:2712397](-)),\n",
       " three_prime_UTR([2711477:2711518](-)),\n",
       " mRNA([2724428:2725543](+)),\n",
       " exon([2724428:2724651](+)),\n",
       " five_prime_UTR([2724428:2724613](+)),\n",
       " exon([2724714:2724772](+)),\n",
       " exon([2724913:2725006](+)),\n",
       " exon([2725114:2725543](+)),\n",
       " three_prime_UTR([2725192:2725543](+)),\n",
       " mRNA([2737034:2739319](+)),\n",
       " exon([2737034:2737325](+)),\n",
       " five_prime_UTR([2737034:2737137](+)),\n",
       " exon([2737404:2737444](+)),\n",
       " exon([2737503:2737825](+)),\n",
       " exon([2737881:2739319](+)),\n",
       " three_prime_UTR([2739230:2739319](+)),\n",
       " mRNA([2772841:2774811](-)),\n",
       " exon([2774456:2774811](-)),\n",
       " five_prime_UTR([2774583:2774811](-)),\n",
       " exon([2774223:2774391](-)),\n",
       " exon([2774048:2774164](-)),\n",
       " exon([2773714:2773995](-)),\n",
       " exon([2772841:2773670](-)),\n",
       " three_prime_UTR([2772841:2773081](-)),\n",
       " mRNA([2782348:2783328](-)),\n",
       " exon([2782960:2783328](-)),\n",
       " five_prime_UTR([2783269:2783328](-)),\n",
       " exon([2782348:2782865](-)),\n",
       " three_prime_UTR([2782348:2782619](-)),\n",
       " mRNA([2799416:2800872](-)),\n",
       " exon([2800526:2800872](-)),\n",
       " five_prime_UTR([2800774:2800872](-)),\n",
       " exon([2799753:2800443](-)),\n",
       " exon([2799416:2799679](-)),\n",
       " three_prime_UTR([2799416:2799555](-)),\n",
       " mRNA([2812544:2815153](+)),\n",
       " exon([2812544:2815153](+)),\n",
       " five_prime_UTR([2812544:2813097](+)),\n",
       " three_prime_UTR([2815041:2815153](+)),\n",
       " mRNA([2823077:2824830](+)),\n",
       " exon([2823077:2823203](+)),\n",
       " five_prime_UTR([2823077:2823128](+)),\n",
       " exon([2823264:2823510](+)),\n",
       " exon([2823596:2823809](+)),\n",
       " exon([2823866:2824830](+)),\n",
       " three_prime_UTR([2824676:2824830](+)),\n",
       " mRNA([2827176:2828962](+)),\n",
       " exon([2827176:2827394](+)),\n",
       " five_prime_UTR([2827176:2827371](+)),\n",
       " exon([2827451:2828962](+)),\n",
       " three_prime_UTR([2828898:2828962](+)),\n",
       " mRNA([2887845:2889191](-)),\n",
       " exon([2888722:2889191](-)),\n",
       " five_prime_UTR([2889154:2889191](-)),\n",
       " exon([2887845:2888654](-)),\n",
       " three_prime_UTR([2887845:2887850](-)),\n",
       " mRNA([2889330:2891204](+)),\n",
       " exon([2889330:2889523](+)),\n",
       " five_prime_UTR([2889330:2889494](+)),\n",
       " exon([2889589:2889618](+)),\n",
       " exon([2889677:2889692](+)),\n",
       " exon([2889747:2889772](+)),\n",
       " exon([2889825:2890144](+)),\n",
       " exon([2890198:2890308](+)),\n",
       " exon([2890360:2890506](+)),\n",
       " exon([2890562:2890620](+)),\n",
       " exon([2890689:2890854](+)),\n",
       " exon([2890938:2891204](+)),\n",
       " three_prime_UTR([2891146:2891204](+)),\n",
       " mRNA([2909365:2910081](-)),\n",
       " exon([2909365:2910081](-)),\n",
       " five_prime_UTR([2909921:2910081](-)),\n",
       " three_prime_UTR([2909365:2909540](-)),\n",
       " mRNA([2907196:2909365](-)),\n",
       " exon([2907196:2909365](-)),\n",
       " five_prime_UTR([2909190:2909365](-)),\n",
       " three_prime_UTR([2907196:2907240](-)),\n",
       " mRNA([2930128:2932214](+)),\n",
       " exon([2930128:2930306](+)),\n",
       " five_prime_UTR([2930128:2930275](+)),\n",
       " exon([2930370:2930761](+)),\n",
       " exon([2930813:2932214](+)),\n",
       " three_prime_UTR([2931855:2932214](+)),\n",
       " mRNA([2986234:2988757](-)),\n",
       " exon([2988077:2988757](-)),\n",
       " five_prime_UTR([2988662:2988757](-)),\n",
       " exon([2986234:2988024](-)),\n",
       " three_prime_UTR([2986234:2986755](-)),\n",
       " mRNA([3018334:3019136](+)),\n",
       " exon([3018334:3019136](+)),\n",
       " five_prime_UTR([3018334:3018422](+)),\n",
       " three_prime_UTR([3019109:3019136](+)),\n",
       " mRNA([3033538:3035208](-)),\n",
       " exon([3033538:3035208](-)),\n",
       " five_prime_UTR([3035162:3035208](-)),\n",
       " three_prime_UTR([3033538:3034091](-)),\n",
       " mRNA([3080290:3082763](-)),\n",
       " exon([3082445:3082763](-)),\n",
       " five_prime_UTR([3082575:3082763](-)),\n",
       " exon([3082282:3082363](-)),\n",
       " exon([3081984:3082192](-)),\n",
       " exon([3081321:3081926](-)),\n",
       " exon([3081210:3081272](-)),\n",
       " exon([3081050:3081154](-)),\n",
       " exon([3080967:3081000](-)),\n",
       " exon([3080290:3080914](-)),\n",
       " three_prime_UTR([3080290:3080394](-)),\n",
       " mRNA([3079803:3080290](-)),\n",
       " exon([3080113:3080290](-)),\n",
       " five_prime_UTR([3080187:3080290](-)),\n",
       " exon([3079803:3080037](-)),\n",
       " three_prime_UTR([3079803:3079895](-)),\n",
       " mRNA([3086871:3088067](+)),\n",
       " exon([3086871:3087296](+)),\n",
       " five_prime_UTR([3086871:3087011](+)),\n",
       " exon([3087396:3088067](+)),\n",
       " three_prime_UTR([3087708:3088067](+)),\n",
       " mRNA([3090244:3091547](-)),\n",
       " exon([3090244:3091547](-)),\n",
       " five_prime_UTR([3091287:3091547](-)),\n",
       " three_prime_UTR([3090244:3090471](-)),\n",
       " mRNA([3107448:3108758](+)),\n",
       " exon([3107448:3108758](+)),\n",
       " five_prime_UTR([3107448:3107869](+)),\n",
       " three_prime_UTR([3108721:3108758](+)),\n",
       " mRNA([43929:44724](-)),\n",
       " exon([44524:44724](-)),\n",
       " five_prime_UTR([44627:44724](-)),\n",
       " exon([43929:44448](-)),\n",
       " three_prime_UTR([43929:43963](-)),\n",
       " mRNA([44825:48337](+)),\n",
       " exon([44825:44986](+)),\n",
       " five_prime_UTR([44825:44892](+)),\n",
       " exon([45066:45212](+)),\n",
       " exon([45259:45286](+)),\n",
       " exon([45337:45373](+)),\n",
       " exon([45432:47841](+)),\n",
       " exon([47911:48337](+)),\n",
       " three_prime_UTR([48085:48337](+)),\n",
       " mRNA([49234:50937](+)),\n",
       " exon([49234:49718](+)),\n",
       " five_prime_UTR([49234:49575](+)),\n",
       " exon([49777:50937](+)),\n",
       " three_prime_UTR([50930:50937](+)),\n",
       " mRNA([56050:57340](-)),\n",
       " exon([57285:57340](-)),\n",
       " five_prime_UTR([57285:57340](-)),\n",
       " exon([56962:57235](-)),\n",
       " five_prime_UTR([57066:57235](-)),\n",
       " exon([56769:56914](-)),\n",
       " exon([56654:56713](-)),\n",
       " exon([56050:56600](-)),\n",
       " three_prime_UTR([56050:56563](-)),\n",
       " mRNA([61708:63066](-)),\n",
       " exon([62937:63066](-)),\n",
       " five_prime_UTR([63064:63066](-)),\n",
       " exon([61708:62801](-)),\n",
       " three_prime_UTR([61708:61902](-)),\n",
       " mRNA([63197:63890](+)),\n",
       " exon([63197:63890](+)),\n",
       " five_prime_UTR([63197:63313](+)),\n",
       " three_prime_UTR([63538:63890](+)),\n",
       " mRNA([73383:75295](+)),\n",
       " exon([73383:73633](+)),\n",
       " five_prime_UTR([73383:73568](+)),\n",
       " exon([73701:73751](+)),\n",
       " exon([73808:73884](+)),\n",
       " exon([73947:74437](+)),\n",
       " exon([74510:75295](+)),\n",
       " three_prime_UTR([75029:75295](+)),\n",
       " mRNA([98842:100600](+)),\n",
       " exon([98842:99043](+)),\n",
       " five_prime_UTR([98842:99041](+)),\n",
       " exon([99128:99190](+)),\n",
       " exon([99256:99405](+)),\n",
       " exon([99466:100600](+)),\n",
       " three_prime_UTR([100330:100600](+)),\n",
       " mRNA([105433:107217](-)),\n",
       " exon([105433:107217](-)),\n",
       " five_prime_UTR([107001:107217](-)),\n",
       " three_prime_UTR([105433:105708](-)),\n",
       " mRNA([108581:110571](+)),\n",
       " exon([108581:109198](+)),\n",
       " five_prime_UTR([108581:108637](+)),\n",
       " exon([109270:109881](+)),\n",
       " exon([109945:110041](+)),\n",
       " exon([110092:110571](+)),\n",
       " three_prime_UTR([110303:110571](+)),\n",
       " mRNA([123805:124644](+)),\n",
       " exon([123805:123948](+)),\n",
       " five_prime_UTR([123805:123866](+)),\n",
       " exon([124024:124285](+)),\n",
       " exon([124355:124462](+)),\n",
       " exon([124510:124644](+)),\n",
       " three_prime_UTR([124564:124644](+)),\n",
       " mRNA([130500:133789](+)),\n",
       " exon([130500:131436](+)),\n",
       " five_prime_UTR([130500:131043](+)),\n",
       " exon([131495:133789](+)),\n",
       " three_prime_UTR([133544:133789](+)),\n",
       " mRNA([143356:144772](+)),\n",
       " exon([143356:143407](+)),\n",
       " five_prime_UTR([143356:143364](+)),\n",
       " exon([143461:143581](+)),\n",
       " exon([143641:143702](+)),\n",
       " exon([143749:143886](+)),\n",
       " exon([143942:144772](+)),\n",
       " three_prime_UTR([144589:144772](+)),\n",
       " mRNA([174150:175094](+)),\n",
       " exon([174150:174310](+)),\n",
       " five_prime_UTR([174150:174193](+)),\n",
       " exon([174426:174671](+)),\n",
       " exon([174753:175094](+)),\n",
       " three_prime_UTR([174910:175094](+)),\n",
       " mRNA([176855:178617](-)),\n",
       " exon([178225:178617](-)),\n",
       " five_prime_UTR([178351:178617](-)),\n",
       " exon([178079:178165](-)),\n",
       " exon([177573:178021](-)),\n",
       " exon([176855:177523](-)),\n",
       " three_prime_UTR([176855:176941](-)),\n",
       " mRNA([178914:180237](-)),\n",
       " exon([180025:180237](-)),\n",
       " five_prime_UTR([180037:180237](-)),\n",
       " exon([179872:179971](-)),\n",
       " exon([179690:179787](-)),\n",
       " exon([179541:179628](-)),\n",
       " exon([179425:179489](-)),\n",
       " exon([179202:179377](-)),\n",
       " exon([178914:179142](-)),\n",
       " three_prime_UTR([178914:179097](-)),\n",
       " mRNA([183622:185094](+)),\n",
       " exon([183622:184166](+)),\n",
       " five_prime_UTR([183622:183958](+)),\n",
       " exon([184232:185094](+)),\n",
       " three_prime_UTR([184675:185094](+)),\n",
       " mRNA([197716:198589](+)),\n",
       " exon([197716:197848](+)),\n",
       " five_prime_UTR([197716:197824](+)),\n",
       " exon([197926:198110](+)),\n",
       " exon([198224:198589](+)),\n",
       " three_prime_UTR([198295:198589](+)),\n",
       " mRNA([207373:209343](-)),\n",
       " exon([208326:209343](-)),\n",
       " five_prime_UTR([209295:209343](-)),\n",
       " exon([207908:208273](-)),\n",
       " exon([207719:207816](-)),\n",
       " exon([207373:207652](-)),\n",
       " three_prime_UTR([207373:207628](-)),\n",
       " mRNA([216598:217959](-)),\n",
       " exon([216598:217959](-)),\n",
       " five_prime_UTR([217901:217959](-)),\n",
       " three_prime_UTR([216598:216770](-)),\n",
       " mRNA([293917:295471](-)),\n",
       " exon([295216:295471](-)),\n",
       " five_prime_UTR([295231:295471](-)),\n",
       " exon([295042:295158](-)),\n",
       " exon([294950:294977](-)),\n",
       " exon([294669:294883](-)),\n",
       " exon([294379:294603](-)),\n",
       " exon([293917:294321](-)),\n",
       " three_prime_UTR([293917:294263](-)),\n",
       " mRNA([315704:316868](+)),\n",
       " exon([315704:315981](+)),\n",
       " five_prime_UTR([315704:315852](+)),\n",
       " exon([316060:316194](+)),\n",
       " exon([316254:316868](+)),\n",
       " three_prime_UTR([316669:316868](+)),\n",
       " mRNA([319473:320708](-)),\n",
       " exon([320246:320708](-)),\n",
       " five_prime_UTR([320352:320708](-)),\n",
       " exon([320092:320187](-)),\n",
       " exon([319473:320028](-)),\n",
       " three_prime_UTR([319473:319644](-)),\n",
       " mRNA([320688:321205](+)),\n",
       " exon([320688:321205](+)),\n",
       " five_prime_UTR([320688:320762](+)),\n",
       " three_prime_UTR([321002:321205](+)),\n",
       " mRNA([321627:323394](+)),\n",
       " exon([321627:321851](+)),\n",
       " five_prime_UTR([321627:321851](+)),\n",
       " exon([321940:322425](+)),\n",
       " five_prime_UTR([321940:321987](+)),\n",
       " exon([322486:322512](+)),\n",
       " exon([322571:322666](+)),\n",
       " exon([322719:322777](+)),\n",
       " exon([322832:323394](+)),\n",
       " three_prime_UTR([322905:323394](+)),\n",
       " mRNA([323393:325183](-)),\n",
       " exon([325083:325183](-)),\n",
       " five_prime_UTR([325083:325183](-)),\n",
       " exon([324900:325026](-)),\n",
       " five_prime_UTR([324981:325026](-)),\n",
       " exon([324799:324843](-)),\n",
       " exon([324670:324748](-)),\n",
       " exon([323393:324624](-)),\n",
       " three_prime_UTR([323393:323738](-)),\n",
       " mRNA([331297:332365](-)),\n",
       " exon([332306:332365](-)),\n",
       " five_prime_UTR([332336:332365](-)),\n",
       " exon([332170:332242](-)),\n",
       " exon([331983:332058](-)),\n",
       " exon([331856:331889](-)),\n",
       " exon([331297:331789](-)),\n",
       " three_prime_UTR([331297:331762](-)),\n",
       " mRNA([336110:336557](-)),\n",
       " exon([336110:336557](-)),\n",
       " five_prime_UTR([336548:336557](-)),\n",
       " three_prime_UTR([336110:336308](-)),\n",
       " mRNA([377856:379544](-)),\n",
       " exon([379111:379544](-)),\n",
       " five_prime_UTR([379326:379544](-)),\n",
       " exon([378890:379054](-)),\n",
       " exon([378550:378837](-)),\n",
       " exon([378191:378494](-)),\n",
       " three_prime_UTR([378191:378314](-)),\n",
       " exon([377856:378112](-)),\n",
       " three_prime_UTR([377856:378112](-)),\n",
       " mRNA([388799:390206](-)),\n",
       " exon([389591:390206](-)),\n",
       " five_prime_UTR([389847:390206](-)),\n",
       " exon([389450:389540](-)),\n",
       " exon([388799:389351](-)),\n",
       " three_prime_UTR([388799:389103](-)),\n",
       " mRNA([409374:410897](-)),\n",
       " exon([410859:410897](-)),\n",
       " five_prime_UTR([410861:410897](-)),\n",
       " exon([410006:410801](-)),\n",
       " exon([409374:409917](-)),\n",
       " three_prime_UTR([409374:409496](-)),\n",
       " mRNA([411066:411490](+)),\n",
       " exon([411066:411490](+)),\n",
       " five_prime_UTR([411066:411098](+)),\n",
       " three_prime_UTR([411329:411490](+)),\n",
       " mRNA([424938:426071](+)),\n",
       " exon([424938:425071](+)),\n",
       " five_prime_UTR([424938:425038](+)),\n",
       " exon([425125:425149](+)),\n",
       " exon([425200:425260](+)),\n",
       " exon([425315:426071](+)),\n",
       " three_prime_UTR([425771:426071](+)),\n",
       " mRNA([426659:429750](+)),\n",
       " exon([426659:426831](+)),\n",
       " five_prime_UTR([426659:426765](+)),\n",
       " exon([427016:427186](+)),\n",
       " exon([427326:427383](+)),\n",
       " exon([427439:427560](+)),\n",
       " exon([427630:428096](+)),\n",
       " exon([428414:429750](+)),\n",
       " three_prime_UTR([429199:429750](+)),\n",
       " mRNA([432790:433987](+)),\n",
       " exon([432790:432879](+)),\n",
       " five_prime_UTR([432790:432829](+)),\n",
       " exon([432947:432981](+)),\n",
       " exon([433034:433401](+)),\n",
       " exon([433457:433619](+)),\n",
       " exon([433675:433987](+)),\n",
       " three_prime_UTR([433869:433987](+)),\n",
       " mRNA([433962:434970](-)),\n",
       " exon([434275:434970](-)),\n",
       " five_prime_UTR([434887:434970](-)),\n",
       " exon([433962:434203](-)),\n",
       " three_prime_UTR([433962:434107](-)),\n",
       " mRNA([450869:451942](+)),\n",
       " exon([450869:450941](+)),\n",
       " five_prime_UTR([450869:450910](+)),\n",
       " exon([451008:451463](+)),\n",
       " exon([451539:451762](+)),\n",
       " exon([451819:451942](+)),\n",
       " three_prime_UTR([451845:451942](+)),\n",
       " mRNA([463696:465696](-)),\n",
       " exon([465563:465696](-)),\n",
       " five_prime_UTR([465609:465696](-)),\n",
       " exon([464773:465515](-)),\n",
       " exon([464665:464721](-)),\n",
       " exon([463696:464559](-)),\n",
       " three_prime_UTR([463696:463996](-)),\n",
       " mRNA([465835:467410](-)),\n",
       " exon([466994:467410](-)),\n",
       " five_prime_UTR([467261:467410](-)),\n",
       " exon([466662:466927](-)),\n",
       " exon([466414:466611](-)),\n",
       " exon([465835:466355](-)),\n",
       " three_prime_UTR([465835:465965](-)),\n",
       " mRNA([477173:479739](+)),\n",
       " exon([477173:477373](+)),\n",
       " five_prime_UTR([477173:477339](+)),\n",
       " exon([477441:477510](+)),\n",
       " exon([477622:477860](+)),\n",
       " exon([477917:479368](+)),\n",
       " exon([479420:479739](+)),\n",
       " three_prime_UTR([479518:479739](+)),\n",
       " mRNA([482966:484138](-)),\n",
       " exon([482966:484138](-)),\n",
       " five_prime_UTR([484052:484138](-)),\n",
       " three_prime_UTR([482966:483008](-)),\n",
       " mRNA([484674:487756](+)),\n",
       " exon([484674:485423](+)),\n",
       " five_prime_UTR([484674:485379](+)),\n",
       " exon([485581:485622](+)),\n",
       " exon([485687:485785](+)),\n",
       " exon([485845:486491](+)),\n",
       " exon([486559:487125](+)),\n",
       " exon([487191:487756](+)),\n",
       " three_prime_UTR([487215:487756](+)),\n",
       " mRNA([498301:499052](-)),\n",
       " exon([498661:499052](-)),\n",
       " five_prime_UTR([498748:499052](-)),\n",
       " exon([498466:498588](-)),\n",
       " exon([498301:498405](-)),\n",
       " three_prime_UTR([498301:498371](-)),\n",
       " mRNA([527017:529118](+)),\n",
       " exon([527017:527530](+)),\n",
       " five_prime_UTR([527017:527022](+)),\n",
       " exon([527581:529118](+)),\n",
       " three_prime_UTR([529023:529118](+)),\n",
       " mRNA([529324:530227](+)),\n",
       " exon([529324:529842](+)),\n",
       " five_prime_UTR([529324:529446](+)),\n",
       " exon([529956:530227](+)),\n",
       " three_prime_UTR([530094:530227](+)),\n",
       " mRNA([530391:532573](+)),\n",
       " exon([530391:530773](+)),\n",
       " five_prime_UTR([530391:530773](+)),\n",
       " exon([530827:531002](+)),\n",
       " five_prime_UTR([530827:530858](+)),\n",
       " exon([531051:531156](+)),\n",
       " exon([531215:531607](+)),\n",
       " exon([531672:531737](+)),\n",
       " exon([531806:531820](+)),\n",
       " exon([531890:532573](+)),\n",
       " three_prime_UTR([532502:532573](+)),\n",
       " mRNA([562078:563344](+)),\n",
       " exon([562078:562364](+)),\n",
       " five_prime_UTR([562078:562195](+)),\n",
       " exon([562430:563344](+)),\n",
       " three_prime_UTR([563290:563344](+)),\n",
       " mRNA([620493:621088](-)),\n",
       " exon([620973:621088](-)),\n",
       " five_prime_UTR([620988:621088](-)),\n",
       " exon([620846:620917](-)),\n",
       " exon([620687:620774](-)),\n",
       " exon([620493:620620](-)),\n",
       " three_prime_UTR([620493:620526](-)),\n",
       " mRNA([627494:628381](-)),\n",
       " exon([628167:628381](-)),\n",
       " five_prime_UTR([628237:628381](-)),\n",
       " exon([627494:628075](-)),\n",
       " three_prime_UTR([627494:627614](-)),\n",
       " mRNA([692534:694056](+)),\n",
       " exon([692534:693013](+)),\n",
       " five_prime_UTR([692534:692674](+)),\n",
       " exon([693080:694056](+)),\n",
       " three_prime_UTR([693824:694056](+)),\n",
       " mRNA([726930:727917](+)),\n",
       " exon([726930:727601](+)),\n",
       " five_prime_UTR([726930:726992](+)),\n",
       " exon([727672:727917](+)),\n",
       " three_prime_UTR([727714:727917](+)),\n",
       " mRNA([751361:753611](+)),\n",
       " exon([751361:751575](+)),\n",
       " five_prime_UTR([751361:751575](+)),\n",
       " exon([751632:753611](+)),\n",
       " five_prime_UTR([751632:751935](+)),\n",
       " three_prime_UTR([753546:753611](+)),\n",
       " mRNA([760453:761452](-)),\n",
       " exon([761381:761452](-)),\n",
       " five_prime_UTR([761381:761452](-)),\n",
       " exon([760817:761255](-)),\n",
       " five_prime_UTR([761024:761255](-)),\n",
       " exon([760453:760770](-)),\n",
       " three_prime_UTR([760453:760509](-)),\n",
       " mRNA([761637:762316](+)),\n",
       " exon([761637:761694](+)),\n",
       " five_prime_UTR([761637:761694](+)),\n",
       " exon([761759:761982](+)),\n",
       " five_prime_UTR([761759:761787](+)),\n",
       " exon([762035:762316](+)),\n",
       " three_prime_UTR([762230:762316](+)),\n",
       " mRNA([785496:787051](+)),\n",
       " exon([785496:785669](+)),\n",
       " five_prime_UTR([785496:785669](+)),\n",
       " exon([785717:786243](+)),\n",
       " five_prime_UTR([785717:785757](+)),\n",
       " exon([786295:787051](+)),\n",
       " three_prime_UTR([786631:787051](+)),\n",
       " mRNA([802812:803949](+)),\n",
       " exon([802812:803515](+)),\n",
       " five_prime_UTR([802812:803320](+)),\n",
       " exon([803577:803949](+)),\n",
       " three_prime_UTR([803874:803949](+)),\n",
       " mRNA([812675:813807](-)),\n",
       " exon([813707:813807](-)),\n",
       " five_prime_UTR([813758:813807](-)),\n",
       " exon([813621:813650](-)),\n",
       " exon([813029:813540](-)),\n",
       " exon([812675:812973](-)),\n",
       " three_prime_UTR([812675:812781](-)),\n",
       " mRNA([847868:850185](+)),\n",
       " exon([847868:848154](+)),\n",
       " five_prime_UTR([847868:847881](+)),\n",
       " exon([848248:850185](+)),\n",
       " mRNA([875419:876720](+)),\n",
       " exon([875419:875694](+)),\n",
       " five_prime_UTR([875419:875448](+)),\n",
       " exon([875797:876145](+)),\n",
       " exon([876229:876720](+)),\n",
       " three_prime_UTR([876616:876720](+)),\n",
       " mRNA([883102:885310](+)),\n",
       " exon([883102:883466](+)),\n",
       " five_prime_UTR([883102:883196](+)),\n",
       " exon([883535:883593](+)),\n",
       " exon([883666:884231](+)),\n",
       " exon([884282:884835](+)),\n",
       " exon([884886:885310](+)),\n",
       " three_prime_UTR([885141:885310](+)),\n",
       " mRNA([894525:895463](+)),\n",
       " exon([894525:894709](+)),\n",
       " five_prime_UTR([894525:894664](+)),\n",
       " exon([894772:894797](+)),\n",
       " exon([894853:895138](+)),\n",
       " exon([895197:895234](+)),\n",
       " exon([895295:895463](+)),\n",
       " three_prime_UTR([895359:895463](+)),\n",
       " mRNA([908979:911365](+)),\n",
       " exon([908979:909120](+)),\n",
       " five_prime_UTR([908979:909074](+)),\n",
       " exon([909185:909849](+)),\n",
       " exon([909913:911365](+)),\n",
       " three_prime_UTR([911111:911365](+)),\n",
       " mRNA([922915:924892](-)),\n",
       " exon([924667:924892](-)),\n",
       " five_prime_UTR([924732:924892](-)),\n",
       " exon([924577:924604](-)),\n",
       " exon([924023:924517](-)),\n",
       " exon([923962:923974](-)),\n",
       " exon([922915:923912](-)),\n",
       " three_prime_UTR([922915:923223](-)),\n",
       " mRNA([933016:934046](-)),\n",
       " exon([933016:934046](-)),\n",
       " five_prime_UTR([933921:934046](-)),\n",
       " three_prime_UTR([933016:933054](-)),\n",
       " mRNA([949862:951394](-)),\n",
       " exon([951027:951394](-)),\n",
       " five_prime_UTR([951240:951394](-)),\n",
       " exon([950627:950966](-)),\n",
       " exon([949862:950491](-)),\n",
       " three_prime_UTR([949862:950179](-)),\n",
       " ...]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records[0].get_all_features()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2853f46-1e5d-455a-8a60-932bc1a7e2d8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
