{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d1c76ffa-43ca-4b5d-9365-90d576a469c6",
   "metadata": {},
   "source": [
    "# Metadata Extraction\n",
    "This notebook scraps the information from JGI html page of the project and returns a Python dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b634635d-ad32-4e89-b049-8b8ca5104bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from bs4 import BeautifulSoup\n",
    "import numpy as np\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ebac1b9-b24f-4c31-ba03-3099a6d22832",
   "metadata": {},
   "source": [
    "Load html result.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d22526a3-e949-4887-9b88-8ab76db54c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../data/raw/Aspergillus whole-genus sequencing.html\", 'r') as f:\n",
    "    contents = f.read()\n",
    "    \n",
    "soup = BeautifulSoup(contents, 'html.parser')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05c02a86-3ec7-4720-9eb8-aec5f10b3bce",
   "metadata": {},
   "source": [
    "Extracting metadata..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "32faa77e-1ada-4a2f-bb6d-9f6495be811f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctr = 0\n",
    "metadata = {}\n",
    "for child in soup.recursiveChildGenerator():\n",
    "    if child.name == 'td':\n",
    "        try:\n",
    "            if child.attrs['data-title'] == \"'Name'\":\n",
    "                name = child.find('div', {'dynamic' : 'result.homelink'})\n",
    "                name = name.text.replace(\"Project:\", \"\").strip()\n",
    "                #print(name)\n",
    "                metadata[ctr] = {'name' : name}\n",
    "            if child.attrs['data-title'] == \"'Resources'\":\n",
    "                resources = child.find('div', {'dynamic' : 'genomePortal'})\n",
    "                resource_status = child.find('span', {'style' : 'font-weight: bold'}).text.strip()\n",
    "                try:\n",
    "                    resource_id = resources.text.strip()\n",
    "                    resource_link = resources.find(\"a\").attrs['href']\n",
    "                except AttributeError:\n",
    "                    resource_id = None\n",
    "                    resource_link = None\n",
    "                #print(resource_id, resource_status, resource_link)\n",
    "                metadata[ctr]['resource_id'] = resource_id\n",
    "                metadata[ctr]['resource_link'] = resource_link\n",
    "            if child.attrs['data-title'] == \"'JGI Info'\":\n",
    "                jgi_info = child.find('a', {'ng-bind-html' : 'result.jgiProjectId'})\n",
    "                project_id = jgi_info.text.strip()\n",
    "                project_link = jgi_info['href']\n",
    "                folder_name = project_link.replace(\"https://genome.jgi.doe.gov/portal/pages/projectStatus.jsf?db=\", \"\")\n",
    "                related_projects = [i.text for i in child.find_all('div', {'dynamic' : 'project'}) if i.text != \"\"]\n",
    "                #print(project_id, related_projects)\n",
    "                metadata[ctr]['project_id'] = project_id\n",
    "                metadata[ctr]['project_link'] = project_link\n",
    "                metadata[ctr]['folder_name'] = folder_name\n",
    "                metadata[ctr]['related_projects'] = related_projects\n",
    "            if child.attrs['data-title'] == \"'Status'\":\n",
    "                status = child.find(\"span\", {\"ng-bind-html\" : \"result.projectStatus\"}).text.strip()\n",
    "                date = child.find(\"span\", {\"ng-bind-html\" : \"result.projectStatusDate\"}).text.strip()\n",
    "                #print(status, date)\n",
    "                metadata[ctr]['status'] = status\n",
    "                metadata[ctr]['release_date'] = date\n",
    "            if child.attrs['data-title'] == \"'Contacts'\":\n",
    "                contacts = child.find('a', {'bo-href' : \"'mailto:'+ stripHtmlTags(proposalPi.email)\"})\n",
    "                contact_name = contacts.text\n",
    "                contact_mail = contacts['href'].replace(\"mailto:\", \"\")\n",
    "                metadata[ctr]['contact_name'] = contact_name\n",
    "                metadata[ctr]['contact_mail'] = contact_mail\n",
    "                ctr = ctr + 1\n",
    "                #print(\"___\")\n",
    "        except KeyError:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9c80fbf-11b5-40f1-9b5e-68afdaefc15a",
   "metadata": {},
   "source": [
    "## Cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "335e2005-47da-4b6c-97a1-7d33355591c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame.from_dict(metadata).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5270d628-74b0-4774-a25b-024c4612dbf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>resource_id</th>\n",
       "      <th>resource_link</th>\n",
       "      <th>project_id</th>\n",
       "      <th>project_link</th>\n",
       "      <th>folder_name</th>\n",
       "      <th>related_projects</th>\n",
       "      <th>status</th>\n",
       "      <th>release_date</th>\n",
       "      <th>contact_name</th>\n",
       "      <th>contact_mail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aspergillus acristatulus CBS 119.55 Annotated ...</td>\n",
       "      <td>Aspacri1</td>\n",
       "      <td>https://mycocosm.jgi.doe.gov/Aspacri1</td>\n",
       "      <td>1052020</td>\n",
       "      <td>https://genome.jgi.doe.gov/portal/pages/projec...</td>\n",
       "      <td>AspacrStandDraft_FD</td>\n",
       "      <td>[SP 1052022, SP 1052021, AP 1052024, AP 105202...</td>\n",
       "      <td>Complete</td>\n",
       "      <td>2014-03-25</td>\n",
       "      <td>Scott E. Baker</td>\n",
       "      <td>scott.baker@pnnl.gov</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aspergillus aculeatinus CBS 121060 Annotated S...</td>\n",
       "      <td>Aspacu1</td>\n",
       "      <td>https://mycocosm.jgi.doe.gov/Aspacu1</td>\n",
       "      <td>1027254</td>\n",
       "      <td>https://genome.jgi.doe.gov/portal/pages/projec...</td>\n",
       "      <td>AspacuStandDraft_FD</td>\n",
       "      <td>[SP 1027256, SP 1027255, AP 1027257]</td>\n",
       "      <td>Complete</td>\n",
       "      <td>2014-12-18</td>\n",
       "      <td>Scott E. Baker</td>\n",
       "      <td>scott.baker@pnnl.gov</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name resource_id  \\\n",
       "0  Aspergillus acristatulus CBS 119.55 Annotated ...    Aspacri1   \n",
       "1  Aspergillus aculeatinus CBS 121060 Annotated S...     Aspacu1   \n",
       "\n",
       "                           resource_link project_id  \\\n",
       "0  https://mycocosm.jgi.doe.gov/Aspacri1    1052020   \n",
       "1   https://mycocosm.jgi.doe.gov/Aspacu1    1027254   \n",
       "\n",
       "                                        project_link          folder_name  \\\n",
       "0  https://genome.jgi.doe.gov/portal/pages/projec...  AspacrStandDraft_FD   \n",
       "1  https://genome.jgi.doe.gov/portal/pages/projec...  AspacuStandDraft_FD   \n",
       "\n",
       "                                    related_projects    status release_date  \\\n",
       "0  [SP 1052022, SP 1052021, AP 1052024, AP 105202...  Complete   2014-03-25   \n",
       "1               [SP 1027256, SP 1027255, AP 1027257]  Complete   2014-12-18   \n",
       "\n",
       "     contact_name          contact_mail  \n",
       "0  Scott E. Baker  scott.baker@pnnl.gov  \n",
       "1  Scott E. Baker  scott.baker@pnnl.gov  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d6ecc883-dc9a-4d36-9b61-de591127f1cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Problem with AspaurMinimDraft_3_FD : Aspaurful1_test\n",
      "   Other id found: Aspaurful1_test --> Aspaurful1\n",
      "\n",
      "Problem with AspcamStandDraft_FD : 2761201767\n",
      "   Other id found: 2761201767 --> Aspcam1\n",
      "\n",
      "Problem with AspcanStandDraft_FD : Aspcan1\n",
      "   Other id found: Aspcan1 --> Aspcand1\n",
      "\n",
      "Problem with AspeucStandDraft_FD : Aspeuc1\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AsphetStandDraft_FD : Asphet1\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AspheyMinimDraft_FD : Asphey1_P5\n",
      "   Other id found: Asphey1_P5 --> Asphey1\n",
      "\n",
      "Problem with AspintMinimDraft_FD : Aspint1\n",
      "   multiple file detected\n",
      "['14097814-Aspint1_AssemblyScaffolds_Repeatmasked.fasta.gz', '14010937-Aspint1_AssemblyScaffolds_Repeatmasked.fasta.gz']\n",
      "\n",
      "Problem with AspisrMinimDraft_FD : Aspisr1_1\n",
      "   Other id found: Aspisr1_1 --> Aspisr1\n",
      "\n",
      "Problem with AsplacStandDraft_FD : Asplac1\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AspmonMinimDraft_FD : Aspmon1\n",
      "   multiple file detected\n",
      "['14097818-Aspmon1_AssemblyScaffolds_Repeatmasked.fasta.gz', '13998657-Aspmon1_AssemblyScaffolds_Repeatmasked.fasta.gz', '14097839-Aspmon1_AssemblyScaffolds_Repeatmasked.fasta.gz']\n",
      "\n",
      "Problem with AspneoMinimDraft_11_FD : Aspneocar1\n",
      "   multiple file detected\n",
      "['14097879-Aspneocar1_AssemblyScaffolds_Repeatmasked.fasta.gz', '13998676-Aspneocar1_AssemblyScaffolds_Repeatmasked.fasta.gz']\n",
      "\n",
      "Problem with AspneoStandDraft_FD : Aspneo1\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AspnovStandDraft_FD : 2761201763\n",
      "   Other id found: 2761201763 --> Aspnov1\n",
      "\n",
      "Problem with AspochStandDraft_FD : 2761201795\n",
      "   Other id found: 2761201795 --> Aspoch1\n",
      "\n",
      "Problem with AsppalMinimDraft_FD : Asppal1\n",
      "   Other id found: Asppal1 --> Asppal1_1\n",
      "\n",
      "Problem with AsppeyMinimDraft_FD : None\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AsppseStandDraft_FD : Asppse1\n",
      "   Other id found: Asppse1 --> Asppsen1\n",
      "\n",
      "Problem with AspsacStandDraft_FD : Aspsac1\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AspsclStandDraft_FD : Aspscl1\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AspsteStandDraft_FD : 2761201819\n",
      "   Other id found: 2761201819 --> Aspste1\n",
      "\n",
      "Problem with AsptepMinimDraft_FD : Asptep1\n",
      "   Other id found: Asptep1 --> Asptep1_1\n",
      "\n",
      "Problem with AspturMinimDraft_FD : None\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AspuvaStandDraft_FD : Aspuva1\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AspvadStandDraft_FD : Aspvad1\n",
      "   WARNING: No assembly found\n",
      "\n",
      "Problem with AspspaMinimDraft_FD : SarstrTest1\n",
      "   Other id found: SarstrTest1 --> Sarstr1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Path of the data\n",
    "data_path = Path(\"../data/raw/\").resolve()\n",
    "\n",
    "# Checking if all assemblies are present\n",
    "for i in df.index:\n",
    "    folder_name = df.loc[i, \"folder_name\"]\n",
    "    resource_id = df.loc[i, \"resource_id\"]\n",
    "    masked_assembly_file = data_path / f\"{folder_name}/Assembly/Mycocosm/Assembly/Assembled_scaffolds__masked_/{resource_id}_AssemblyScaffolds_Repeatmasked.fasta.gz\"\n",
    "    if masked_assembly_file.is_file():\n",
    "        df.loc[i, \"masked_assemblies\"] = masked_assembly_file\n",
    "        df.loc[i, \"assembly_id_alias\"] = resource_id\n",
    "    else:\n",
    "        print(f\"Problem with {folder_name} : {resource_id}\")\n",
    "        test_path = data_path / f\"{folder_name}/Assembly/Mycocosm/Assembly/Assembled_scaffolds__masked_/\"\n",
    "        assembly_files = [i for i in test_path.glob(\"*.fasta.gz\")]\n",
    "        if len(assembly_files) > 1:\n",
    "            print(\"   multiple file detected\")\n",
    "            print([i.name for i in assembly_files])\n",
    "            assembly_files = assembly_files\n",
    "        else:\n",
    "            try:\n",
    "                correct_id = assembly_files[0].name.replace(\"_AssemblyScaffolds_Repeatmasked.fasta.gz\", \"\")\n",
    "                print(f\"   Other id found: {resource_id} --> {correct_id}\")\n",
    "                df.loc[i, \"assembly_id_alias\"] = correct_id\n",
    "            except:\n",
    "                print(\"   WARNING: No assembly found\")\n",
    "                assembly_files = np.nan\n",
    "        df.at[i, \"masked_assemblies\"] = assembly_files\n",
    "        print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0ea6d30c-2da7-4f2c-9057-802bea5b2c27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>resource_id</th>\n",
       "      <th>resource_link</th>\n",
       "      <th>project_id</th>\n",
       "      <th>project_link</th>\n",
       "      <th>folder_name</th>\n",
       "      <th>related_projects</th>\n",
       "      <th>status</th>\n",
       "      <th>release_date</th>\n",
       "      <th>contact_name</th>\n",
       "      <th>contact_mail</th>\n",
       "      <th>masked_assemblies</th>\n",
       "      <th>assembly_id_alias</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aspergillus acristatulus CBS 119.55 Annotated ...</td>\n",
       "      <td>Aspacri1</td>\n",
       "      <td>https://mycocosm.jgi.doe.gov/Aspacri1</td>\n",
       "      <td>1052020</td>\n",
       "      <td>https://genome.jgi.doe.gov/portal/pages/projec...</td>\n",
       "      <td>AspacrStandDraft_FD</td>\n",
       "      <td>[SP 1052022, SP 1052021, AP 1052024, AP 105202...</td>\n",
       "      <td>Complete</td>\n",
       "      <td>2014-03-25</td>\n",
       "      <td>Scott E. Baker</td>\n",
       "      <td>scott.baker@pnnl.gov</td>\n",
       "      <td>/datadrive/matin_other_projects/jgi_aspergillu...</td>\n",
       "      <td>Aspacri1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aspergillus aculeatinus CBS 121060 Annotated S...</td>\n",
       "      <td>Aspacu1</td>\n",
       "      <td>https://mycocosm.jgi.doe.gov/Aspacu1</td>\n",
       "      <td>1027254</td>\n",
       "      <td>https://genome.jgi.doe.gov/portal/pages/projec...</td>\n",
       "      <td>AspacuStandDraft_FD</td>\n",
       "      <td>[SP 1027256, SP 1027255, AP 1027257]</td>\n",
       "      <td>Complete</td>\n",
       "      <td>2014-12-18</td>\n",
       "      <td>Scott E. Baker</td>\n",
       "      <td>scott.baker@pnnl.gov</td>\n",
       "      <td>/datadrive/matin_other_projects/jgi_aspergillu...</td>\n",
       "      <td>Aspacu1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name resource_id  \\\n",
       "0  Aspergillus acristatulus CBS 119.55 Annotated ...    Aspacri1   \n",
       "1  Aspergillus aculeatinus CBS 121060 Annotated S...     Aspacu1   \n",
       "\n",
       "                           resource_link project_id  \\\n",
       "0  https://mycocosm.jgi.doe.gov/Aspacri1    1052020   \n",
       "1   https://mycocosm.jgi.doe.gov/Aspacu1    1027254   \n",
       "\n",
       "                                        project_link          folder_name  \\\n",
       "0  https://genome.jgi.doe.gov/portal/pages/projec...  AspacrStandDraft_FD   \n",
       "1  https://genome.jgi.doe.gov/portal/pages/projec...  AspacuStandDraft_FD   \n",
       "\n",
       "                                    related_projects    status release_date  \\\n",
       "0  [SP 1052022, SP 1052021, AP 1052024, AP 105202...  Complete   2014-03-25   \n",
       "1               [SP 1027256, SP 1027255, AP 1027257]  Complete   2014-12-18   \n",
       "\n",
       "     contact_name          contact_mail  \\\n",
       "0  Scott E. Baker  scott.baker@pnnl.gov   \n",
       "1  Scott E. Baker  scott.baker@pnnl.gov   \n",
       "\n",
       "                                   masked_assemblies assembly_id_alias  \n",
       "0  /datadrive/matin_other_projects/jgi_aspergillu...          Aspacri1  \n",
       "1  /datadrive/matin_other_projects/jgi_aspergillu...           Aspacu1  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "654240d8-8cee-4625-a976-753927e0db38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Problem with AspacuStandDraft_FD : Aspacu1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspaurMinimDraft_3_FD : Aspaurful1_test\n",
      "   Other id found: Aspaurful1_test --> Aspaurful1\n",
      "\n",
      "Problem with AspbruStandDraft_FD : Aspbru1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspcamStandDraft_FD : 2761201767\n",
      "   Other id found: 2761201767 --> Aspcam1\n",
      "\n",
      "Problem with AspcanStandDraft_FD : Aspcan1\n",
      "   Other id found: Aspcan1 --> Aspcand1\n",
      "\n",
      "Problem with AspcosStandDraft_FD : Aspcos1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspellStandDraft_FD : Aspell1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspeucStandDraft_FD : Aspeuc1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspfijStandDraft_FD : Aspfij1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AsphetStandDraft_FD : Asphet1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AsphetMinimDraft_FD : Asphethal1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspheyMinimDraft_FD : Asphey1_P5\n",
      "   Other id found: Asphey1_P5 --> Asphey1\n",
      "\n",
      "Problem with AsphomStandDraft_FD : Asphom1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspibeStandDraft_FD : Aspibe1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspimpStandDraft_FD : Aspimp1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspindStandDraft_2_FD : Aspind2_1\n",
      "   Other id found: Aspind2_1 --> Aspind2\n",
      "\n",
      "Problem with AspindStandDraft_FD : Aspind1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspintMinimDraft_FD : Aspint1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspisrMinimDraft_FD : Aspisr1_1\n",
      "   Other id found: Aspisr1_1 --> Aspisr1\n",
      "\n",
      "Problem with AspjapStandDraft_FD : Aspjap1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AsplacStandDraft_FD : Asplac1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspmonMinimDraft_FD : Aspmon1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspmulStandDraft_FD : Aspmul1_1\n",
      "   Other id found: Aspmul1_1 --> Aspmul1\n",
      "\n",
      "Problem with AspnavStandDraft_FD : Aspnav1_1\n",
      "   Other id found: Aspnav1_1 --> Aspnav1\n",
      "\n",
      "Problem with AspneoMinimDraft_11_FD : Aspneocar1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspneoMinimDraft_4_FD : Aspneog1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspneoStandDraft_FD : Aspneo1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspnovStandDraft_FD : 2761201763\n",
      "   Other id found: 2761201763 --> Aspnov1\n",
      "\n",
      "Problem with AspochStandDraft_FD : 2761201795\n",
      "   Other id found: 2761201795 --> Aspoch1\n",
      "\n",
      "Problem with AspoliStandDraft_FD : Aspoli1_1\n",
      "   Other id found: Aspoli1_1 --> Aspoli1\n",
      "\n",
      "Problem with AspparMinimDraft_FD : Aspparac1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AsppetStandDraft_FD : Asppet1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AsppeyMinimDraft_FD : None\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AsppipStandDraft_FD : Asppip1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AsppseStandDraft_FD : Asppse1\n",
      "   Other id found: Asppse1 --> Asppsen1\n",
      "\n",
      "Problem with AsprecStandDraft_FD : Asprec1_1\n",
      "   Other id found: Asprec1_1 --> Asprec1\n",
      "\n",
      "Problem with AspsacStandDraft_FD : Aspsac1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspsclStandDraft_2_FD : Aspscle1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspsclStandDraft_FD : Aspscl1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspsimStandDraft_FD : Aspsim1_1\n",
      "   Other id found: Aspsim1_1 --> Aspsim1\n",
      "\n",
      "Problem with AspspeStandDraft_FD : Aspspe1_1\n",
      "   Other id found: Aspspe1_1 --> Aspspe1\n",
      "\n",
      "Problem with AspsteStandDraft_4_FD : Aspstel1_1\n",
      "   Other id found: Aspstel1_1 --> Aspstel1\n",
      "\n",
      "Problem with AspsteStandDraft_2_FD : Aspstec1_1\n",
      "   Other id found: Aspstec1_1 --> Aspstec1\n",
      "\n",
      "Problem with AspsteStandDraft_FD : 2761201819\n",
      "   Other id found: 2761201819 --> Aspste1\n",
      "\n",
      "Problem with AspsulStandDraft_FD : Aspsul1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AsptriMinimDraft_FD : Asptri1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspturMinimDraft_FD : None\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspundStandDraft_FD : Aspund1_1\n",
      "   Other id found: Aspund1_1 --> Aspund1\n",
      "\n",
      "Problem with AspungStandDraft_FD : Aspung1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspuniMinimDraft_FD : Aspuni1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspuvaStandDraft_FD : Aspuva1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspvadStandDraft_FD : Aspvad1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspvenStandDraft_FD : Aspven1\n",
      "   multiple file detected\n",
      "\n",
      "Problem with AspvioStandDraft_FD : Aspvio1\n",
      "   WARNING: No annotation found\n",
      "\n",
      "Problem with AspspaMinimDraft_FD : SarstrTest1\n",
      "   Other id found: SarstrTest1 --> Sarstr1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Checking if all annotations are present\n",
    "for i in df.index:\n",
    "    folder_name = df.loc[i, \"folder_name\"]\n",
    "    resource_id = df.loc[i, \"resource_id\"]\n",
    "    gene_annotation_path = data_path / f\"{folder_name}/Annotation/Mycocosm/Annotation/Filtered_Models___best__/Genes/\"\n",
    "    gene_annotations = [i for i in gene_annotation_path.glob(\"*.gff3.gz\")]\n",
    "    if len(gene_annotations) > 1:\n",
    "        print(f\"Problem with {folder_name} : {resource_id}\")\n",
    "        print(\"   multiple file detected\")\n",
    "        print(\"\")\n",
    "        gff3 = gene_annotations\n",
    "    elif len(gene_annotations) == 1:\n",
    "        gff3 = gene_annotations[0]\n",
    "        correct_id = gff3.name.split(\"_\")[0]\n",
    "        if resource_id != correct_id:\n",
    "            print(f\"Problem with {folder_name} : {resource_id}\")\n",
    "            print(f\"   Other id found: {resource_id} --> {correct_id}\")\n",
    "            print(\"\")\n",
    "            df.loc[i, \"annotation_id_alias\"] = correct_id\n",
    "        else:\n",
    "            df.loc[i, \"annotation_id_alias\"] = resource_id\n",
    "    elif len(gene_annotations) == 0:\n",
    "        print(f\"Problem with {folder_name} : {resource_id}\")\n",
    "        print(\"   WARNING: No annotation found\")\n",
    "        print(\"\")\n",
    "        gff3 = np.nan\n",
    "    df.at[i, \"annotations\"] = gff3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1419d5c2-aaee-42b7-b9d6-27e74f1b21bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>resource_id</th>\n",
       "      <th>resource_link</th>\n",
       "      <th>project_id</th>\n",
       "      <th>project_link</th>\n",
       "      <th>folder_name</th>\n",
       "      <th>related_projects</th>\n",
       "      <th>status</th>\n",
       "      <th>release_date</th>\n",
       "      <th>contact_name</th>\n",
       "      <th>contact_mail</th>\n",
       "      <th>masked_assemblies</th>\n",
       "      <th>assembly_id_alias</th>\n",
       "      <th>annotation_id_alias</th>\n",
       "      <th>annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aspergillus acristatulus CBS 119.55 Annotated ...</td>\n",
       "      <td>Aspacri1</td>\n",
       "      <td>https://mycocosm.jgi.doe.gov/Aspacri1</td>\n",
       "      <td>1052020</td>\n",
       "      <td>https://genome.jgi.doe.gov/portal/pages/projec...</td>\n",
       "      <td>AspacrStandDraft_FD</td>\n",
       "      <td>[SP 1052022, SP 1052021, AP 1052024, AP 105202...</td>\n",
       "      <td>Complete</td>\n",
       "      <td>2014-03-25</td>\n",
       "      <td>Scott E. Baker</td>\n",
       "      <td>scott.baker@pnnl.gov</td>\n",
       "      <td>/datadrive/matin_other_projects/jgi_aspergillu...</td>\n",
       "      <td>Aspacri1</td>\n",
       "      <td>Aspacri1</td>\n",
       "      <td>/datadrive/matin_other_projects/jgi_aspergillu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aspergillus aculeatinus CBS 121060 Annotated S...</td>\n",
       "      <td>Aspacu1</td>\n",
       "      <td>https://mycocosm.jgi.doe.gov/Aspacu1</td>\n",
       "      <td>1027254</td>\n",
       "      <td>https://genome.jgi.doe.gov/portal/pages/projec...</td>\n",
       "      <td>AspacuStandDraft_FD</td>\n",
       "      <td>[SP 1027256, SP 1027255, AP 1027257]</td>\n",
       "      <td>Complete</td>\n",
       "      <td>2014-12-18</td>\n",
       "      <td>Scott E. Baker</td>\n",
       "      <td>scott.baker@pnnl.gov</td>\n",
       "      <td>/datadrive/matin_other_projects/jgi_aspergillu...</td>\n",
       "      <td>Aspacu1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name resource_id  \\\n",
       "0  Aspergillus acristatulus CBS 119.55 Annotated ...    Aspacri1   \n",
       "1  Aspergillus aculeatinus CBS 121060 Annotated S...     Aspacu1   \n",
       "\n",
       "                           resource_link project_id  \\\n",
       "0  https://mycocosm.jgi.doe.gov/Aspacri1    1052020   \n",
       "1   https://mycocosm.jgi.doe.gov/Aspacu1    1027254   \n",
       "\n",
       "                                        project_link          folder_name  \\\n",
       "0  https://genome.jgi.doe.gov/portal/pages/projec...  AspacrStandDraft_FD   \n",
       "1  https://genome.jgi.doe.gov/portal/pages/projec...  AspacuStandDraft_FD   \n",
       "\n",
       "                                    related_projects    status release_date  \\\n",
       "0  [SP 1052022, SP 1052021, AP 1052024, AP 105202...  Complete   2014-03-25   \n",
       "1               [SP 1027256, SP 1027255, AP 1027257]  Complete   2014-12-18   \n",
       "\n",
       "     contact_name          contact_mail  \\\n",
       "0  Scott E. Baker  scott.baker@pnnl.gov   \n",
       "1  Scott E. Baker  scott.baker@pnnl.gov   \n",
       "\n",
       "                                   masked_assemblies assembly_id_alias  \\\n",
       "0  /datadrive/matin_other_projects/jgi_aspergillu...          Aspacri1   \n",
       "1  /datadrive/matin_other_projects/jgi_aspergillu...           Aspacu1   \n",
       "\n",
       "  annotation_id_alias                                        annotations  \n",
       "0            Aspacri1  /datadrive/matin_other_projects/jgi_aspergillu...  \n",
       "1                 NaN                                                NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "58131172-0b98-42ab-b6fd-4c457963b2b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "245"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many samples in total?\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b06979d7-274f-4a30-9ea8-27312473d78a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total annotated samples: 245. There are 220 good assemblies, 15 samples with multiple assemblies, and 10 samples with no assembly\n",
      "Total annotated samples: 245.There are 210 good annotation, 13 samples with multiple annotation, and 22 samples with no annotation\n"
     ]
    }
   ],
   "source": [
    "def evaluate_jgi(df, column_name):\n",
    "    null = []\n",
    "    multiple = []\n",
    "    good = []\n",
    "    for i in df.index:\n",
    "        item = df.loc[i, column_name]\n",
    "        if type(item) == pathlib.PosixPath:\n",
    "            assert item.is_file()\n",
    "            good.append(i)\n",
    "        elif type(item) == list:\n",
    "            for link in item:\n",
    "                assert link.is_file()\n",
    "            multiple.append(i)\n",
    "        else:\n",
    "            null.append(i)\n",
    "    return good, multiple, null\n",
    "\n",
    "# How many assemblies can we process?\n",
    "good_assembly, multiple_assembly, null_assembly = evaluate_jgi(df, \"masked_assemblies\")\n",
    "print(f\"Total annotated samples: {len(good_assembly + multiple_assembly + null_assembly )}. There are {len(good_assembly)} good assemblies, {len(multiple_assembly)} samples with multiple assemblies, and {len(null_assembly)} samples with no assembly\")\n",
    "\n",
    "# How many annotations can we process?\n",
    "good_annot, multiple_annot, null_annot = evaluate_jgi(df, \"annotations\")\n",
    "print(f\"Total annotated samples: {len(good_annot + multiple_annot + null_annot )}.There are {len(good_annot)} good annotation, {len(multiple_annot)} samples with multiple annotation, and {len(null_annot)} samples with no annotation\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3b722a8c-d528-4623-8024-369aad6b62be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have 198 samples with both single annotation and assembly\n"
     ]
    }
   ],
   "source": [
    "# How many samples have good assembly and annotation?\n",
    "good_intersect = set(good_annot).intersection(set(good_assembly))\n",
    "print(f\"We have {len(good_intersect)} samples with both single annotation and assembly\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "eafb213b-d89d-466f-bbc3-478a5be31600",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have 223 samples which is not null\n"
     ]
    }
   ],
   "source": [
    "# How many samples have good assembly and annotation?\n",
    "not_null_intersect = set(good_annot + multiple_annot).intersection(set(good_assembly + multiple_assembly))\n",
    "print(f\"We have {len(not_null_intersect)} samples which is not null\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e50aa56e-180d-48e9-a054-9648a5c93e31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 Aspind2_1 Aspind2_1 Aspind2\n",
      "123 Aspmul1_1 Aspmul1_1 Aspmul1\n",
      "127 Aspnav1_1 Aspnav1_1 Aspnav1\n",
      "147 Aspoli1_1 Aspoli1_1 Aspoli1\n",
      "178 Asprec1_1 Asprec1_1 Asprec1\n",
      "193 Aspsim1_1 Aspsim1_1 Aspsim1\n",
      "195 Aspspe1_1 Aspspe1_1 Aspspe1\n",
      "198 Aspstel1_1 Aspstel1_1 Aspstel1\n",
      "200 Aspstec1_1 Aspstec1_1 Aspstec1\n",
      "226 Aspund1_1 Aspund1_1 Aspund1\n"
     ]
    }
   ],
   "source": [
    "clean_df = df.loc[list(good_intersect), :]\n",
    "for i in clean_df.index:\n",
    "    row = df.loc[i, :]\n",
    "    if row.resource_id == row.assembly_id_alias == row.annotation_id_alias:\n",
    "        pass\n",
    "    else:\n",
    "        print(i, row.resource_id, row.assembly_id_alias, row.annotation_id_alias)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d663df3c-61ce-460f-a77a-9125b0d53ed9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "198"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(clean_df.project_id.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "cfa2186c-dec9-44df-872d-67e1be4a013d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "198"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(clean_df.annotation_id_alias.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "45a4c6d1-9ced-4218-8c57-f7fec53ff9fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "table_out = Path(\"../tables\")\n",
    "table_out.mkdir()\n",
    "clean_df.to_csv(\"../tables/clean_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "151abc36-1e42-49c9-85df-fa7f0261d5cf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
